---
title: "Benchmarking bRF and LASSO-D3S"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(dplyr)
library(ggpubr)
library(clusterProfiler)
library(org.At.tair.db)
library(patchwork)
library(visNetwork)
library(igraph)
source('inference_functions/bRF.R')
source('inference_functions/LASSO-D3S.R')
source('inference_functions/evaluateNetwork.R')
source('inference_functions/MSE.R')
theme_set(theme_pubr())
```

# Context and data

This document benchmarks the bRF and LASSO-D3S functions for integrative GRN inference.

> If you are interested in using those functions on your data, and especially on your personal computer, we recommend that you go to the tutorial Rmd file instead of this one. This benchmark is intended to be used only to reproduce the results of the article, and and requires very intensive computing.

Those functions infer the regulatory pathways of *Arabidopsis thaliana*'s roots in response to nitrate (N) induction from [Varala et al., 2018](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE97500).

They use as inputs the expression profiles of N-responsive genes and TFBS information. 
Prior TFBS information was built by searching in the promoters of the N-responsive genes the PWM of the N-responsive regulators. 

## Data import

### Expression data

Import of the expression data and the N-responsive genes and regulators :

```{r}
load('rdata/inference_input_N_response_varala.rdata')
genes <- input_data$grouped_genes; length(genes)
tfs <- input_data$grouped_regressors; length(tfs)
counts <- input_data$counts; dim(counts)
```

### TFBS data

Presence or absence of the TFs motifs in the targets promoter :

```{r}
load("rdata/pwm_occurrences_N_response_varala.rdata")
dim(pwm_occurrence)
```


# Benchmark

## Benchmark funtion 

```{r}
infer_networks <- function(alphas, densities, 
                         counts, genes, tfs, 
                         nCores = 50, nrep = 3){
  edges <- list()
    
  for(alpha in alphas){ # exploring PWM integration strength
    for(rep in 1:nrep){ # exploring inherent variability
      
      mat_rf <- bRF_inference(counts, genes, tfs, nTrees = 1000,
                              alpha = alpha, prior_strength = 2,
                              pwm_occurrence = pwm_occurrence, 
                              nCores = nCores, 
                              importance = "%IncMSE")
      
      mat_lasso <- LASSO.D3S_inference(counts, genes, tfs,
                                       alpha = alpha, N = 50,
                                       robustness = 0.1,int_pwm_noise = 0,
                                       pwm_occurrence = pwm_occurrence,
                                       nCores = nCores)
      
      for(density in densities){ # exploring importance threshold stringency
        
        edges[[paste0("bRF_", as.character(alpha),  '_', density, '_', rep)]] <- 
          bRF_network(mat_rf, density = density, pwm_occurrence, genes, tfs)
        
        edges[[paste0("LASSO-D3S_", as.character(alpha),  '_', density, '_', rep)]] <-
          LASSO.D3S_network(mat_lasso, density = density, pwm_occurrence, genes, tfs)
      }
    }
  }
  return(edges)
}
```

## Actual benchmark 

```{r}
ALPHAS <- seq(0,1, by = 0.1)
DENSITIES <- c(0.005, 0.01, 0.025, 0.05, 0.075) 

set.seed(1213)
edges <- infer_networks(alphas = ALPHAS, densities = DENSITIES, 
             counts = counts, genes = genes, tfs = tfs, 
             nCores = 11, nrep = 3)

save(edges, file = "results/main_results.rdata")
```


# Plotting results

## PWM Support

```{r, fig.width=15}
load("results/main_results.rdata")
color_palette = c("#88002D", "#C12131", "#EC5D2F", "#FE945C", "#FFC08E" )
prettyZero <- function(l){
    max.decimals = max(nchar(str_extract(l, "\\.[0-9]+")), na.rm = T)-1
    lnew = formatC(l, replace.zero = T, zero.print = "0",
        digits = max.decimals, format = "f", preserve.width=T)
    return(lnew)
}


settings <- c("method", "alpha", "density", "rep")

# number of edges per network
nrows <-
  data.frame(alpha_rep = names(unlist(lapply(edges, FUN = nrow))),
             n_edges = unlist(lapply(edges, FUN = nrow)))
nrows[, settings] <-
  str_split_fixed(nrows$alpha_rep, '_', length(settings))

edges_num <- lapply(edges, function(df)
  df[sapply(df, is.numeric)])
pwm_support <-
  data.frame(alpha_rep = names(unlist(lapply(edges_num, FUN = nrow))),
             pwm = unlist(lapply(edges_num, FUN = colMeans)))
pwm_support[, settings] <-
  str_split_fixed(pwm_support$alpha_rep, '_', length(settings))

pwm_support_plot <- pwm_support %>%
  left_join(nrows, by = settings) %>%
  mutate(alpha = as.numeric(alpha),
         density_label = paste(density, ':', n_edges, 'edges')) %>%
  ggplot(aes(color = density_label, x = alpha, y = pwm)) +
  ggh4x::facet_nested_wrap(vars(method), ncol = 8, nest_line = T) +
  geom_point() +
  geom_smooth(aes(fill = density_label), alpha = 0.1) +
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.position = "none",
    legend.text = element_text(size = 15),
    axis.text = element_text(size = 12)
  ) +
  xlab(expression(alpha)) + ylab("Mean PWM score in selected edges") +
  ggtitle("Average PWM support of inferred edges") +
  guides(color = guide_legend(nrow = 1, byrow = TRUE),
         fill = guide_legend(nrow = 1, byrow = TRUE)) +
  ylab(expression(paste("mean(", pi[tr], ")"))) +
  scale_x_continuous(labels = prettyZero) +
  geom_hline(yintercept = pwm_support_baseline, color = "grey", linewidth=2) +
  annotate("text", x=0.5, y=pwm_support_baseline+0.025, label="Baseline", color = "darkgrey", size=5)+
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)
pwm_support_plot

```


## Validation against ConnecTF

### Running the valdiation


```{r, fig.width=15}
val_conn <-
  evaluate_networks(
    edges,
    validation = c("TARGET", "CHIPSeq", "DAPSeq"),
    nCores = 10,
    input_genes = genes,
    input_tfs = tfs
  )
val_conn[, settings] <-
  str_split_fixed(val_conn$network_name, '_', length(settings))

```



### Computing baseline precision and MSE (random networks)



```{r}
N = 100000
baseline <- data.frame("from" = sample(tfs, size = N, replace = T), 
                       "to" = sample(genes, size = N, replace = T))
base <- evaluate_network(
    baseline,
    validation = c("TARGET", "CHIPSeq", "DAPSeq"),
    input_genes = genes,
    input_tfs = tfs
  )

precision_baseline <- base$tpr
mse_baseline <- get_MSE_baseline(counts, genes[1:1000]);mse_baseline
pwm_support_baseline <- mean(pwm_imputed[cbind(baseline$to, baseline$from)])
```




### Plotting precision and recall analyses

```{r, fig.width=7, fig.height=8}
load("results/val_conn.rdata")
val_connecTF <-  val_conn %>%
  dplyr::select(-network_name) %>%
  left_join(nrows, by = settings) %>%
  mutate(alpha = as.numeric(alpha)) %>%
  mutate(density_label = paste(density, ':', n_edges, 'edges')) %>%
  ggplot(aes(color = density_label, x = alpha, y = precision)) +
  ggh4x::facet_nested_wrap(vars(method), ncol = 8, nest_line = T) + geom_point() +
  geom_smooth(aes(fill = density_label), alpha = 0.1) +
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_text(size = 12),
    legend.position = 'none'
  ) +
  geom_hline(yintercept = precision_baseline, color = "grey", linewidth=2) +
  annotate("text", x=0.5, y=precision_baseline+0.01, label="Baseline", color = "darkgrey", size=5)+
  xlab(expression(alpha)) + ylab("Precision") +
  ggtitle("Precision against ConnecTF") +
  scale_x_continuous(labels = prettyZero) +
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)
val_connecTF


val_connecTF_recall <-  val_conn %>%
  dplyr::select(-network_name) %>%
  left_join(nrows, by = settings) %>%
  mutate(alpha = as.numeric(alpha)) %>%
  mutate(density_label = paste(density, ':', n_edges, 'edges')) %>%
  ggplot(aes(color = density_label, x = alpha, y = recall)) +
  ggh4x::facet_nested_wrap(vars(method), ncol = 8, nest_line = T) + geom_point() +
  geom_smooth(aes(fill = density_label), alpha = 0.1) +
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_text(size = 12),
    legend.position = 'none'
  ) +
  scale_x_continuous(labels = prettyZero) +
  xlab(expression(alpha)) + ylab("Recall") +
  ggtitle("Recall against ConnecTF") +
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)
val_connecTF_recall
```



# Prediction error 

## Computing prediction error on test expression data

```{r}

mses <- list()
for(network in names(edges)){
  mses[[network]] <- get_MSE(edges[[network]], method = str_split_fixed(network, '_', 2)[,1], nCores=10)
}
save(mses, file = "results/mse.rdata")

```

```{r}
load("results/mse.rdata")
settings <- c("method", "alpha", "density", "rep")
mse <- data.frame(alpha_rep = names(unlist(mses)),
                mmmse = unlist(mses))
mse[, settings] <- str_split_fixed(mse$alpha_rep,'_', length(settings))

mse_plot <- mse %>%  
  left_join(nrows, by = settings) %>%
  mutate(alpha = as.numeric(alpha)) %>%
  mutate(density_label = paste(density, ':', n_edges, 'edges')) %>%
  ggplot(aes(color = density_label, x = alpha, y = mmmse)) +
  ggh4x::facet_nested_wrap(vars(method), ncol = 8, nest_line = T) + geom_point() +
  geom_smooth(aes(fill = density_label), alpha = 0.1) +
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_text(size = 12),
    legend.position = 'top'
  ) +
  geom_hline(yintercept = mse_baseline, color = "grey", linewidth=2) +
  annotate("text", x=0.5, y=mse_baseline+0.2, label="Baseline", color = "darkgrey", size=5)+
  scale_x_continuous(labels = prettyZero) +
  xlab(expression(alpha))+ ylab("Log(mean(MSE))") + 
                 ggtitle("Log of average MSE") +
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)
```


## Main figure

```{r, fig.width=14, fig.height=10}
figure <- pwm_support_plot + mse_plot + val_connecTF + val_connecTF_recall + 
  patchwork::plot_annotation(tag_levels = "a") + plot_layout(ncol=2, guides = "collect"); figure

ggexport(figure, filename = "results/figure3.pdf", width = 14, height = 10)
```



## Intersection of predicted edges between the two methods

### Common edges and common hubs

```{r}

intersections <- list()
intersection_hubs <- list()
top_hubs <- 20


for(alpha in ALPHAS){ # exploring PWM integration strength
  for(rep in 1:3){ # exploring inherent variability
    for(density in DENSITIES){ # exploring importance threshold stringency
      
      net1 <- edges[[paste0("bRF_", as.character(alpha),  '_', 
                     density, '_', rep)]]
      net2 <- edges[[paste0("LASSO-D3S_", as.character(alpha),  
                     '_', density, '_',  rep)]]
      
      # intersection of top degree regulators from network 1 and 2
      intersection_hubs[[paste0(as.character(alpha), '_', density, '_', rep)]] <-
        length(intersect(names(sort(-table(net1$from))[1:top_hubs]), 
                  names(sort(-table(net2$from))[1:top_hubs])))/top_hubs
      
      intersections[[paste0(as.character(alpha), '_', density, '_', rep)]] <- 
        inner_join(net1, net2, by = c("from", "to", "pwm"))
    }
  }
}


settings_intersection <- c("alpha", "density", "rep")
# number of edges per network
nrows_int <- data.frame(alpha_rep = names(unlist(lapply(intersections, FUN = nrow))),
                n_edges = unlist(lapply(intersections, FUN = nrow)))
nrows_int[, settings_intersection] <- str_split_fixed(nrows_int$alpha_rep,'_', length(settings_intersection))
n_edges <- nrows_int %>%
  mutate(alpha = as.numeric(alpha), 
         total_edges = as.numeric(nrows$n_edges[seq_along(nrows$n_edges) %%2 == 0])) %>%
  ggplot(aes(color = density, x = alpha, y = n_edges/total_edges)) +
  geom_point() + 
  geom_smooth(aes(fill = density), alpha = 0.1) +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), legend.position = 'none',
        legend.text = element_text(size = 15), axis.text = element_text(size = 12)) +
  xlab(expression(alpha)) + ylab("Fraction of common edges") + 
  ggtitle("Fraction of edges shared \nby bRF and LASSO-3DS") + 
  scale_x_continuous(labels = prettyZero)+
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)



```



## Plotting the results


```{r, fig.width=10}


get_nTF <- function(net){
  return(mean(table(net$to)))
}

get_TF <- function(net){
  return(length(unique(net$from)))
}

topo <- data.frame(alpha_rep = names(edges),
                n_TFs = unlist(lapply(edges, FUN = get_nTF)),
                TFs = unlist(lapply(edges, FUN = get_TF)))

topo[, settings] <- str_split_fixed(topo$alpha_rep,'_', length(settings))

nTFs <- topo %>%  
  left_join(nrows, by = settings) %>%
  mutate(alpha = as.numeric(alpha)) %>%
  mutate(density_label = paste(density, ':', n_edges, 'edges')) %>%
  ggplot(aes(color = density_label, x = alpha, y = n_TFs)) +
  ggh4x::facet_nested_wrap(vars(method), ncol = 8, nest_line = T) + geom_point() +
  geom_smooth(aes(fill = density_label), alpha = 0.1) +
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_text(size = 12),
    legend.position = 'none'
  ) +
  scale_x_continuous(labels = prettyZero) +
  xlab(expression(alpha)) + ylab("Average number of regulator per gene") + 
                 ggtitle("Average number of regulator per gene") +
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)



nRegs <- topo %>%  
  left_join(nrows, by = settings) %>%
  mutate(alpha = as.numeric(alpha)) %>%
  mutate(density_label = paste(density, ':', n_edges, 'edges')) %>%
  ggplot(aes(color = density_label, x = alpha, y = TFs)) +
  ggh4x::facet_nested_wrap(vars(method), ncol = 8, nest_line = T) + geom_point() +
  geom_smooth(aes(fill = density_label), alpha = 0.1) +
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_text(size = 12),
    legend.position = 'top'
  ) +
  scale_x_continuous(labels = prettyZero) +
  xlab(expression(alpha))+ ylab("Number of distinct regulators") + 
                 ggtitle("Number of distinct regulators") + 
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)


figure <- n_edges+nTFs+nRegs + plot_layout(widths = c(1,1, 1), guides = "collect") + plot_annotation(tag_levels = 'a');figure
ggexport(figure, filename = "results/figure4.pdf", width = 16, height = 6)
```

# Study of Nitrate-related genes in the inferred GRNs

## Degree of important regulators of Nitrate nutrition


```{r}
load("rdata/arabidopsis_annotation.rdata")
n_genes_favorites <- c("LBD37", "LBD38", "LBD39", "TGA1", "TGA4",
                       "HRS1", "HHO2", "HHO3", "HHO4", "DIV1" )

pwm_imputed <- pwm_occurrence
pwm_imputed[is.na(pwm_imputed)] <- 0.5

network_data <- function(graph){
  data <- toVisNetworkData(graph)
  degree <- igraph::degree(graph)
  data$nodes$degree <- degree[match(data$nodes$id, names(degree))]
  data$nodes$group <- ifelse(data$nodes$id %in% tfs, "Regulator", "Target Gene")
  data$nodes[, colnames(annot)] <- annot[match(data$nodes$id, 
            rownames(annot)), ]
  return(data$nodes)
}

get_nodes_degree <- function(graph){
  data <- toVisNetworkData(graph)
  degree <- igraph::degree(graph)
  data$nodes$degree <- degree[match(data$nodes$id, names(degree))]
  return(data$nodes)
}
networks_N_genes <- function(edges_list, n_genes, nCores = 11){
  
  registerDoParallel(cores = nCores)
  message(paste("\nParallel network N validation Using", foreach::getDoParWorkers(), "cores."))
  "%dopar%" <- foreach::"%dopar%"
  tic()
  suppressPackageStartupMessages(
    result.validation <-
      doRNG::"%dorng%"(foreach::foreach(network_name = names(edges_list), .combine = rbind),
                       {
                         net<- graph_from_data_frame(edges_list[[network_name]], directed = TRUE)
                         nodes <- network_data(net)
                         nodes$pwm_status = ifelse(nodes$group != "Regulator", "target", NA)
                         nodes[is.na(nodes$pwm_status),]$pwm_status =ifelse(
                           colMeans(pwm_imputed[,nodes[is.na(nodes$pwm_status),]$id]) != 0.5, "known", "unknown" )
                         nodes$degree_rank <- rank(nodes$degree)/nrow(nodes)
                         res <- nodes[str_detect(nodes$label, paste(n_genes, collapse = '|')),
                                      c("label", "degree", "degree_rank", "id")]
                         res$known_pwm <- ifelse(colMeans(pwm_imputed[,res$id]) != 0.5, "PWM available", "PWM missing")
                         res <- rbind.data.frame(res, c(" Mean", mean(nodes[nodes$pwm_status == "known",]$degree),
                                                        mean(nodes[nodes$pwm_status == "known",]$degree_rank), NA, "PWM available"))
                         res <- rbind.data.frame(res, c(" Mean", mean(nodes[nodes$pwm_status == "unknown",]$degree),
                                                        mean(nodes[nodes$pwm_status == "unknown",]$degree_rank), NA, "PWM missing"))
                         
                         rownames(res) <- NULL
                         res$network_name <- network_name
                         res
                       })
  )
  toc()
  attr(result.validation, "rng") <- NULL # It contains the whole sequence of RNG seeds
  return(result.validation)
}

N_rank_fav <- networks_N_genes(edges, n_genes_favorites, nCores = 11)
N_rank_fav[,settings] <- str_split_fixed(N_rank_fav$network_name,'_',length(settings))

n_genes_favored <- N_rank_fav %>%
  mutate(`Mean value` = str_detect(label, "mean"))%>% 
  filter(density %in% c(0.005, 0.025, 0.05)) %>%
  dplyr::select(-network_name) %>%
  mutate(alpha = as.numeric(alpha),
         method = str_replace(method, "Biased Random Forests", "bRF"),
         method = str_replace(method, "Differentially shrinked lasso SS", "LASSO-D3S"),
         density = paste0("D=", density)) %>%
  ggplot(aes(color = label, x = alpha, fill = label, y = as.numeric(degree_rank))) +
  ggh4x::facet_nested_wrap(vars(method, known_pwm, density), ncol = 12, nest_line = T) + 
  theme_pubr() +
  geom_smooth(aes(fill = label, linetype = `Mean value`), alpha = 0.15) +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), 
        legend.text = element_text(size = 15), axis.text = element_text(size = 12), 
        legend.position = 'left') +
  xlab(expression(alpha)) + ylab("Relative rank in degree") + 
  ggtitle("Degree rank of N relevant TFs in the inferred GRNs") + 
  scale_x_continuous(labels = prettyZero, breaks =c(0,0.5,  1))+
  scale_linetype_discrete(guide = "none")+ylim(0.4,1)+
  guides(color=guide_legend(nrow=10,byrow=TRUE),
         fill=guide_legend(nrow=10,byrow=TRUE))+
  scale_color_discrete(name = "Regulator") +
  scale_fill_discrete(name = "Regulator") +
  scale_y_continuous(labels = prettyZero);n_genes_favored
```

## Ontology enrichment along network degree

```{r}

N_go <- c("GO:0006807", "GO:0051171", "GO:0051173", "GO:0022622")


convert_from_agi <- function(ids){
  x <- org.At.tair.db::org.At.tairENTREZID
  mapped_genes <- AnnotationDbi::mappedkeys(x)
  xx <- as.list(x[mapped_genes])
  return(unlist(xx[as.vector(ids)]))
}


networks_N_GSEA <- function(edges_list, N_go, nCores = 1){
  
  registerDoParallel(cores = nCores)
  message(paste("\nParallel network N validation Using", foreach::getDoParWorkers(), "cores."))
  "%dopar%" <- foreach::"%dopar%"
  tic()
  suppressPackageStartupMessages(
    result.validation <-
      doRNG::"%dorng%"(foreach::foreach(network_name = names(edges_list), .combine = rbind),
                       {
                         net<- graph_from_data_frame(edges_list[[network_name]], directed = TRUE)
                         nodes <- get_nodes_degree(net)
                        
                         rownames(nodes) <- nodes$id
                         nodes[names(convert_from_agi(nodes$id)), "entrez"] <- 
                           convert_from_agi(nodes$id)
                         
                         degrees <- sort(setNames(nodes$degree, nodes$entrez), decreasing = T)
                         
                         gse <- gseGO(geneList=degrees, 
                                      ont ="BP", 
                                      minGSSize = 3, 
                                      maxGSSize = 800, nPerm =1000,
                                      pvalueCutoff = 1, 
                                      verbose = TRUE, scoreType = "pos",
                                      OrgDb = org.At.tair.db::org.At.tair.db, 
                                      pAdjustMethod = "BH")
                         res <- gse@result[gse@result$ID %in% N_go,c("Description", "enrichmentScore", "p.adjust")]
                         if(nrow(res)>1){
                           res$network_name <- network_name
                           res
                         }
                         
                       })
  )
  toc()
  attr(result.validation, "rng") <- NULL # It contains the whole sequence of RNG seeds
  return(result.validation)
}


N_gsea <- networks_N_GSEA(edges, N_go, nCores = 1)
N_gsea[,settings] <- str_split_fixed(N_gsea$network_name,'_',length(settings))

save(N_gsea, file = "results/N_gsea.rdata")

n_genes_gsea <- N_gsea %>%
  filter(density %in% c(0.005, 0.025, 0.05)) %>%
  dplyr::select(-network_name) %>%
  mutate(alpha = as.numeric(alpha),
         density = paste0("D=", density),
         Pathway = str_replace(Description, "nitrogen compound", "N"),
         Pathway = str_replace(Pathway, "positive regulation of N metabolic process",
                               "positive regulation of\nN metabolic process"),
         Pathway = str_replace(Pathway, "regulation of N metabolic process",
                               "regulation of N metabolic\nprocess")) %>%
  ggplot(aes(color = Pathway, x = as.numeric(alpha), y = enrichmentScore, 
             fill = Pathway, group = Pathway)) +
  ggh4x::facet_nested_wrap(vars(method, density), ncol = 10, nest_line = T) + 
   theme_pubr() +guides(color=guide_legend(nrow=5,byrow=TRUE))+
  geom_smooth(aes(col = Pathway), alpha=0.3, span = 2) +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), 
        legend.text = element_text(size = 12), axis.text = element_text(size = 12), 
        legend.position = 'left',legend.spacing.y = unit(0.15, 'cm')) + ylim(c(0.5,1))+
  xlab(expression(alpha)) + ylab("Enrichment score") + 
  ggtitle("GSEA of N pathways in the inferred GRNs") + 
  scale_x_continuous(labels = prettyZero, breaks =c(0,0.5,  1));n_genes_gsea

```

## Regultors of NRT2.1


```{r, fig.width=10}
describe_node <- function(graph, node){
  targets <- neighborhood(graph, nodes = node, mode = "out")[[1]]$name
  regulators <- neighborhood(graph, nodes = node, mode = "in")[[1]]$name
  return(list(targets = setdiff(targets, node), 
              regulators = setdiff(regulators, node)))
}

NRT_regs <- function(edges_list, agi, nCores = 32){
  registerDoParallel(cores = nCores)
  message(paste("\nParallel network N validation Using", foreach::getDoParWorkers(), "cores."))
  "%dopar%" <- foreach::"%dopar%"
  tic()
  suppressPackageStartupMessages(
    result.validation <-
      doRNG::"%dorng%"(foreach::foreach(network_name = names(edges_list), .combine = rbind),
                       {
                         net<- graph_from_data_frame(edges_list[[network_name]], directed = TRUE)
                         nodes <- get_nodes_degree(net)
                         if(agi %in% nodes$id){
                           regs <- describe_node(net,agi)$regulators
                           if(!is.null(regs)){
                             # <- get_gene_information(regs, organism = "Arabidopsis thaliana")
                             
                             infos <- annot[match(regs, rownames(annot)),]
                             res <- data.frame(regulators  = infos$label, 
                                               knwon_pwm = colMeans(data.frame(pwm_imputed[,c(regs)]) != 0.5) )
                             res$network_name <- network_name
                             res
                           }
                         }
                       })
  )
  toc()
  attr(result.validation, "rng") <- NULL # It contains the whole sequence of RNG seeds
  return(result.validation)
}

# AT5G60770
# AT5G50200
# AT1G08090

nrt2.1 <- NRT_regs(edges, "AT1G08090", nCores = 11)
nrt2.1[,settings] <- str_split_fixed(nrt2.1$network_name,'_',length(settings))

nrt2.1_plot <- nrt2.1 %>%
  filter(regulators %in% c("HHO3", "HHO6", "NLP7", "NPL1", "LBD38", 
                           "HHO1", "HRS1", "LBD39", "BT1", "BT2")) %>%
  dplyr::select(-network_name) %>%
  filter(density > 0.01) %>%
  mutate(alpha = as.numeric(alpha),
         method = str_replace(method, "Biased Random Forests", "bRF"),
         method = str_replace(method, "Differentially shrinked lasso SS", "LASSO-D3S"),
         density = paste0("D=", density),
         knwon_pwm = ifelse(knwon_pwm==1, " PWM in promoter", "No PWM")) %>%
  ggplot(aes(fill = regulators, x = alpha, y = regulators)) +
  ggh4x::facet_nested_wrap(vars(method, knwon_pwm, density),drop = F,ncol = 6, nest_line = T) + 
  geom_tile(color = "grey", alpha = 0.4) + theme_pubr() +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), 
        legend.text = element_text(size = 15), axis.text = element_text(size = 12), 
        legend.position = 'none') +
  xlab(expression(alpha)) + ylab("") + 
  ggtitle("Regulators of NRT2.1 in the inferred GRNs") + 
  scale_x_continuous(labels = prettyZero, breaks =c(0,0.5,  1));nrt2.1_plot


fig <- n_genes_favored / (n_genes_gsea + nrt2.1_plot) + 
  plot_annotation(tag_levels = 'a') + plot_layout(heights = c(1,1))

ggexport(fig,  filename = "results/figure5.pdf", width = 18, height = 13)

```



### Biological validation of common edges between the two methods


```{r, fig.width=17, fig.height=8}


hubs <- data.frame(network_name = names(intersection_hubs),
                   common_hubs = unlist(intersection_hubs))

hubs[, settings_intersection] <- str_split_fixed(hubs$network_name,'_', length(settings_intersection))
hubs_shared <- hubs %>%
  mutate(alpha = as.numeric(alpha)) %>%
  ggplot(aes(color = density, x = alpha, y = common_hubs)) +
  geom_point() + 
  geom_smooth(aes(fill = density), alpha = 0.1) +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), legend.position = 'none',
        legend.text = element_text(size = 15), axis.text = element_text(size = 12)) +
  xlab(expression(alpha)) + ylab("Fraction of hubs") + 
  ggtitle("Fraction of hubs (top 20) shared") + 
  scale_x_continuous(labels = prettyZero)+
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)


val_inter <- evaluate_networks(intersections, validation = c("TARGET", "CHIPSeq", "DAPSeq"), 
                              nCores = 10, input_genes = genes, input_tfs = tfs)
val_inter[,settings_intersection] <- str_split_fixed(val_inter$network_name,'_',length(settings_intersection))
val_inter_plot <- val_inter %>%
  mutate(alpha = as.numeric(alpha), 
         total_edges = as.numeric(nrows$n_edges[seq_along(nrows$n_edges) %%2 == 0])) %>%
  ggplot(aes(color = density, x = alpha, y = precision)) +
  geom_point() + 
  geom_smooth(aes(fill = density), alpha = 0.1) +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), legend.position = 'top',
        legend.text = element_text(size = 15), axis.text = element_text(size = 12)) +
  xlab(expression(alpha)) + ylab("Precision") + 
  ggtitle("Precision of edges shared") + 
  scale_x_continuous(labels = prettyZero)+
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)


val_inter_plot_recall <- val_inter %>%
  mutate(alpha = as.numeric(alpha), 
         total_edges = as.numeric(nrows$n_edges[seq_along(nrows$n_edges) %%2 == 0])) %>%
  ggplot(aes(color = density, x = alpha, y = recall)) +
  geom_point() + 
  geom_smooth(aes(fill = density), alpha = 0.1) +
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), legend.position = 'none',
        legend.text = element_text(size = 15), axis.text = element_text(size = 12)) +
  xlab(expression(alpha)) + ylab("Recall") + 
  ggtitle("Recall of edges shared") + 
  scale_x_continuous(labels = prettyZero)+
  scale_color_manual(name = "Network density", values = color_palette) +
  scale_fill_manual(name = "Network density", values = color_palette)

val_inter_plot + val_inter_plot_recall + hubs_shared
```