---
title: "Gene specific tuning of integrative GRN inference"
output: 
  html_document:
    df_print: paged
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: hide
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = F, message = F, fig.width = 10)

source('inference_functions/weightedRF.R')
source('inference_functions/weightedLASSO.R')
source('inference_functions/evaluateNetwork.R')
source('inference_functions/data_integration_optimization.R')

library(ggplot2)
library(tidyverse)
library(ggpubr)
library(patchwork)
library(ggVennDiagram)
library(ComplexHeatmap)
library(clusterProfiler)
library(circlize)
```

This document demonstrates the use of the weightedRF and weightedLASSO functions for integrative GRN inference.

It shows how to select one optimal alpha value per gene, in order to apply data integration only to the target genes for which expression prediction error is reduced by using PWM prior information, significantly more than in a baseline model.

# Data import

Import of the expression data and binding motifs data for the nitrate-responsive genes and regulators :

```{r}
load('rdata/inference_input_N_response_varala.rdata')
genes <- input_data$grouped_genes; length(genes)
tfs <- input_data$grouped_regressors; length(tfs)
counts <- input_data$counts; dim(counts)
load("rdata/pwm_occurrences_N_response_varala.rdata")
dim(pwm_occurrence)

ALPHAS <- seq(0,1, by = 0.1)
```


# Getting the MSE and importance metrics of inferred GRNs

For $alpha$ from 0 to 1 with a step of 0.1, infers 100 integrative models on 

+ true data

+ a shuffled baseline.

The following code chunks are very long to run.

## For weightedRF

```{r, eval=FALSE}
nCores = 45
mats <- list()
nrep <- 100
for(alpha in ALPHAS){ # exploring PWM integration strength
  for(rep in 1:nrep){ # exploring inherent variability
    
    mat_rf <- weightedRF_inference(counts, genes, tfs, nTrees = 2000,
                            alpha = alpha,
                            pwm_occurrence = pwm_occurrence,
                            nCores = nCores,
                            importance = "%IncMSE")
    
    mat_rf_perm <- weightedRF_inference(counts, genes, tfs, nTrees = 2000,
                            alpha = alpha, tf_expression_permutation = TRUE,
                            pwm_occurrence = pwm_occurrence,
                            nCores = nCores,
                            importance = "%IncMSE")
    
    mats[[paste0("bRF_", as.character(alpha),  '_trueData_', rep)]] <- mat_rf
    mats[[paste0("bRF_", as.character(alpha),  '_shuffled_', rep)]] <- mat_rf_perm
    
  }
}
save(mats, file = "results/100_permutations_bRF_importances_inf.rdata")

# thresholds the regulatory weights at a certain density to build GRNs
edges <- list()
lmses <- data.frame(row.names = genes)
densities <- c(0.005, 0.01,0.05)
for(name in names(mats)){ 
      for(density in densities){# exploring importance threshold stringency
        edges[[paste0(name, '_', density)]] <-
        weightedRF_network(mats[[name]], density = density, pwm_occurrence, genes, tfs)
        lmses[,name] <- mats[[name]]["mse",]
      }
}
save(edges, file = "results/100_permutations_bRF_edges_inf.rdata")
save(lmses, file = "results/brf_perumtations_mse_all_genes_predict.rdata")

# validation of GRN edges against DAP-Seq
settings <- c("model", "alpha", "dataset", "rep", "density")
val_dap <-
  evaluate_networks(
    edges,
    validation = c("DAPSeq"),
    nCores = 35,
    input_genes = genes,
    input_tfs = tfs
  )
val_dap[, settings] <-
  str_split_fixed(val_dap$network_name, '_', length(settings))
save(val_dap, file = "results/100_permutations_rf_validation_dap_inf.rdata")
```

## For weightedLASSO

```{r, eval=FALSE}
nCores = 40
# mats <- list()
nrep <- 50
for(alpha in c(1)){ # exploring PWM integration strength
  for(rep in 1:nrep){ # exploring inherent variability
    
    mat_lasso <- weightedLASSO_inference(counts = counts, genes = genes, tfs = tfs,
                                     alpha = alpha, N = 50, 
                                     tf_expression_permutation = FALSE,
                                     int_pwm_noise = 0, mda_type = "shuffle",
                                     pwm_occurrence = pwm_occurrence,
                                     nCores = nCores)
    
    mat_lasso_perm <- weightedLASSO_inference(counts, genes, tfs,
                                     alpha = alpha, N = 50, 
                                     tf_expression_permutation = TRUE,
                                     int_pwm_noise = 0, mda_type = "shuffle",
                                     pwm_occurrence = pwm_occurrence,
                                     nCores = nCores)
    
    mats[[paste0("LASSO_", as.character(alpha),  '_trueData_', rep)]] <- mat_lasso
    mats[[paste0("LASSO_", as.character(alpha),  '_shuffled_', rep)]] <- mat_lasso_perm
    
  }
}

save(mats, file = "results/100_permutations_lasso_mda_shuffle.rdata")

# thresholds the regulatory weights at certain densities to build GRNs
edges <- list()
lmses <- data.frame(row.names = genes)
densities <- c(0.005, 0.01,0.05)
for(name in names(mats)){ 
      for(density in densities){# exploring importance threshold stringency
        edges[[paste0(name, '_', density)]] <-
        weightedLASSO_network(mats[[name]], density = density, pwm_occurrence, 
                          genes, tfs, decreasing = TRUE)
        
        lmses[,name] <- mats[[name]]["mse",]
      }
}

save(edges, file = "results/100_permutations_lasso_edges.rdata")
save(lmses, file = "results/lasso_perumtations_mse_all_genes_test.rdata")


# validation of GRN edges against DAP-Seq
settings <- c("model", "alpha", "dataset", "rep", "density")
val_dap <-
  evaluate_networks(
    edges,
    validation = c("DAPSeq"),
    nCores = 35,
    input_genes = genes,
    input_tfs = tfs
  )
val_dap[, settings] <-
  str_split_fixed(val_dap$network_name, '_', length(settings))
save(val_dap, file = "results/100_permutations_lasso_validation_dap.rdata")

```


# Gene-level optimisation of TFBM information


## Plotting EDI behaviour for all genes

How does the effective data integration vary when alpha increases?

### In weightedRF

```{r, fig.height=20}

col_fun = colorRamp2(c(1, 201), hcl_palette = "Blue-Red 3")

# loads the mats object (importance of TF-target interactions)
load("results/100_permutations_bRF_importances_inf.rdata")


# getting mean EDI values for all genes as a function of alpha
edi_rf <- mcsapply(genes, draw_gene_effective_integration, mats = mats,
                          return=T, mc.cores=40)

edi <- t(as.data.frame(edi_rf[3,]))
colnames(edi) = ALPHAS
edi <- na.omit(edi)
ha = HeatmapAnnotation(
    alpha = anno_simple(colnames(edi)),
    annotation_name_side = "left")

rf <- Heatmap(edi, col = col_fun,
        show_column_names = T,
         width = ncol(edi)*unit(10, "mm"), 
      height = nrow(edi)*unit(0.2, "mm"),
      name = "EDI",
        cluster_columns = F, show_row_names = F)

png("results/supp_figures/EDI_all_genes_weightedRF.png", 
    res = 300, height = 1700+2200, width = 2000)
rf
dev.off()
```

# In weightedLASSO

```{r, fig.height=20}
# loads the mats object (importance of TF-target interactions)
load("results/100_permutations_lasso_mda_shuffle.rdata")

# getting mean EDI values for all genes as a function of alpha
edi_lasso <- mcsapply(genes, draw_gene_effective_integration, mats = mats,
                          return=T, mc.cores=40)

edi <- t(as.data.frame(edi_lasso[3,]))
colnames(edi) = ALPHAS
edi <- na.omit(edi)
lasso <- Heatmap(edi, 
        col = col_fun, show_column_names = T,
         width = ncol(edi)*unit(10, "mm"), 
      height = nrow(edi)*unit(0.2, "mm"),
      name = "EDI",
        cluster_columns = F, show_row_names = F)
png("results/supp_figures/EDI_all_genes_weightedLASSO.png", 
    res = 300, height = 1700+2200, width = 2000)
lasso
dev.off()
```


TFBM support of globally optimized networks :

```{r}
# loads lasso edges
load(file = "results/100_permutations_lasso_edges.rdata")
edges_num <- lapply(edges, function(df) df[sapply(df, is.numeric)])
d <- data.frame(settings = names(unlist(lapply(edges_num, FUN = nrow))),
                pwm = unlist(lapply(edges_num, FUN = colMeans)))
d[c("model", "alpha", "dataset", "rep", "density")] <- str_split_fixed(d$settings, '_', 5)
d_lasso <- d


# loads rf edges
load(file = "results/100_permutations_bRF_edges_inf.rdata")
edges_num <- lapply(edges, function(df) df[sapply(df, is.numeric)])
d <- data.frame(settings = names(unlist(lapply(edges_num, FUN = nrow))),
                pwm = unlist(lapply(edges_num, FUN = colMeans)))
d[c("model", "alpha", "dataset", "rep", "density")] <- str_split_fixed(d$settings, '_', 5)
d <- rbind.data.frame(d, d_lasso)

# plots the TFBM support for both models :
pwm_support <- d %>%
  mutate(alpha = as.numeric(alpha),
         model = str_replace(model, "bRF", "weightedRF"),
         model = str_replace(model, "LASSO", "weightedLASSO"),
         density = paste("D =", density)) %>%
  ggplot(aes(color = dataset, x = alpha, y = pwm)) +
  geom_point() + 
  geom_smooth() +
  theme_bw()+
  ggh4x::facet_nested_wrap(vars(model, density), ncol =3, nest_line = T) + 
  theme(strip.background = element_blank(), axis.title.x = element_text(size = 22),
        title = element_text(size = 16), strip.text = element_text(size = 16), 
        legend.text = element_text(size = 15), axis.text = element_text(size = 12)) +
  xlab(expression(alpha)) + ylab("Mean of TFBM scores") + 
  scale_color_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData")))+
  ggtitle("Mean of TFBM scores in inferred GRN edges") ; pwm_support
ggsave(pwm_support, file = "results/supp_figures/TFBM_support.pdf", width = 11, height = 8)
```


## Some gene examples to illustrate our gene-specific optimization of alpha


### In weightedRF

Plots the results for the proposed gene-specific criterion for some gene examples :

```{r, fig.width=9, fig.height=8}
# loads the MSE data for weightRF
load("results/brf_perumtations_mse_all_genes_predict.rdata")

# loads the importances
load("results/100_permutations_bRF_importances_inf.rdata")

gene_no_benefit <- "AT1G30270"
gene_benefit_optimum <- "AT1G14720"
gene_benefit <- "AT5G48970"

n <- draw_gene_effective_integration(gene_no_benefit, mats = mats, prior = 1) +
  draw_gene_mse(gene_no_benefit, lmses = lmses)+
  get_opt_alpha_per_gene(gene_no_benefit, mats = mats, lmses = lmses)+
  plot_annotation(title = gene_no_benefit) 
o <- draw_gene_effective_integration(gene_benefit_optimum, mats = mats, prior = 1) +
  draw_gene_mse(gene_benefit_optimum, lmses = lmses)+theme(legend.position = "none")+
  get_opt_alpha_per_gene(gene_benefit_optimum, mats = mats, lmses = lmses)+
  plot_annotation(title = gene_benefit_optimum) 
p <- draw_gene_effective_integration(gene_benefit, mats = mats, prior = 1) +
  draw_gene_mse(gene_benefit, lmses = lmses)+theme(legend.position = "none")+
  get_opt_alpha_per_gene(gene_benefit, mats = mats, lmses = lmses)+
  plot_annotation(title = gene_benefit)


figure <- n/o/p;figure
ggexport(figure, filename = "results/gene_examples_weightedRF.pdf", width = 10, height = 9)

```


Some more figures illustrating the approach on several gene cases for the ISMB 2023 talk : 


```{r}
# mse increase (or AT5G64660)
compromise <- get_opt_alpha_per_gene("AT1G14720", mats = mats, lmses = lmses)+
  labs(subtitle = "AT1G14720")+
  get_opt_alpha_per_gene("AT5G16650", mats = mats, lmses = lmses)+
  labs(subtitle = "AT5G16650")
 

# mse decrease
ideal <- get_opt_alpha_per_gene("AT1G70790", mats = mats, lmses = lmses)+
  theme(legend.position = "right") + 
  labs(subtitle = "AT1G70790")


# mse increase_bad
bad <- get_opt_alpha_per_gene("AT1G56230", mats = mats, lmses = lmses) +
  labs(subtitle = "AT1G56230")+ 
  get_opt_alpha_per_gene("AT2G29990", mats = mats, lmses = lmses)+
  labs(subtitle = "AT2G29990")


ggexport(ideal, filename = "results/ismb_talk/wRF_ideal.pdf", width = 6, height = 5)
ggexport(compromise, filename = "results/ismb_talk/wRF_compromise.pdf", width = 10, height = 5)
ggexport(bad, filename = "results/ismb_talk/wRF_bad.pdf", width = 10, height = 5)


eff_data <- draw_gene_effective_integration("AT1G14720", mats = mats)+
   theme(legend.position = "right") 

ggexport(eff_data, filename = "results/ismb_talk/wRF_eff_data_int.pdf", 
         width = 6, height = 5)

mse <- draw_gene_mse("AT1G70790", lmses = lmses)+
   theme(legend.position = "right") 

ggexport(mse, filename = "results/ismb_talk/wRF_mse.pdf", 
         width = 6, height = 5)

```






Returns the optimal value of alpha for all nitrate responsive genes :

```{r, eval=F}
# value of alpha per genes:
alphas_rf <- mcsapply(genes, get_opt_alpha_per_gene,
                          return_alpha=T, mc.cores=34)
save(alphas_rf, file = "results/alpha_per_gene_weighted_RF_mean_true_shuff_sd.rdata")

# doing the same but with the value of alpha minimising the mse:
alphas_rf <- mcsapply(genes, get_opt_alpha_per_gene, metric = "min",
                          return_alpha=T, mc.cores=34)
save(alphas_rf, file = "results/alpha_per_gene_weighted_RF_min.rdata")
```


### In weightedLASSO:


```{r fig.height=10, fig.width=10}
# loads mse and importances for weightedLASSO
load("results/lasso_perumtations_mse_all_genes_test.rdata")
load("results/100_permutations_lasso_mda_shuffle.rdata")

n <- draw_gene_effective_integration(gene_no_benefit, mats = mats, prior = 1) +
  draw_gene_mse(gene_no_benefit, lmses = lmses)+
  get_opt_alpha_per_gene(gene_no_benefit, mats = mats, lmses = lmses)+
  plot_annotation(title = gene_no_benefit) 
o <- draw_gene_effective_integration(gene_benefit_optimum, mats = mats, prior = 1) +
  draw_gene_mse(gene_benefit_optimum, lmses = lmses)+theme(legend.position = "none")+
  get_opt_alpha_per_gene(gene_benefit_optimum, mats = mats, lmses = lmses)+
  plot_annotation(title = gene_benefit_optimum) 
p <- draw_gene_effective_integration(gene_benefit, mats = mats, prior = 1) +
  draw_gene_mse(gene_benefit, lmses = lmses)+theme(legend.position = "none")+
  get_opt_alpha_per_gene(gene_benefit, mats = mats, lmses = lmses)+
  plot_annotation(title = gene_benefit)


n/o/p
figure <- n/o/p
ggexport(figure, filename = "results/gene_examples_weightedLASSO.pdf", width = 10, height = 9)
```



```{r, eval=FALSE}
# value of alpha per genes:
alphas_lasso <- mcsapply(genes, get_opt_alpha_per_gene, mats = mats, 
                         lmses = lmses,
                          return_alpha=T, mc.cores=34)
save(alphas_lasso, file = "results/alpha_per_gene_weighted_LASSO_test_mean_true_shuff_sd.rdata")

# doing the same but with the value of alpha minimising the mse:
alphas_lasso <- mcsapply(genes, get_opt_alpha_per_gene, 
                          return_alpha=T, metric = "min", mc.cores=34)
save(alphas_lasso, file = "results/alpha_per_gene_weighted_LASSO_min.rdata")
```


## Distributions of optimal alphas

For maximal divergence from permutation criterion :

```{r}
# optimal values of alpha :
load("results/alpha_per_gene_weighted_LASSO_test_mean_true_shuff_sd.rdata")
load("results/alpha_per_gene_weighted_RF_mean_true_shuff_sd.rdata")

hists <- data.frame(weightedLASSO = alphas_lasso, 
                    weightedRF = alphas_rf) %>%
  gather(key = "model", value = "alpha") %>%
ggplot(aes(x = alpha, fill = model)) +
  geom_histogram()+
  facet_wrap(~model)+
  theme(axis.title.y = element_blank(), legend.position = "none")+
  scale_fill_manual( values = c("#C55A11","#E67F87"))+theme_bw()+
  theme(strip.background = element_blank(),
        legend.position = "none")+
  xlab(expression(alpha))+ ylab("Number of genes")
ggexport(hists, filename  = "results/alpha_histograms.pdf", height = 3, width = 7)
```



# Plotting MSE behaviour for all genes

Depending on their class (optimal alpha different from 0 or not).

## weightedRF

```{r, fig.height=10}
load("results/brf_perumtations_mse_all_genes_predict.rdata")
load(file = "results/alpha_per_gene_weighted_RF_mean_true_shuff_sd.rdata")

pos_class <- names(alphas_rf[alphas_rf!=0])

mse <- lmses[str_detect(colnames(lmses), "trueData")]
for(alpha in seq(0,1, by = 0.1)){
  mse[,paste("alpha",alpha)] <- rowMeans(mse[,str_detect(colnames(mse), paste0("RF_",as.character(alpha), "_"))])
}
mse <- as.matrix(mse[str_detect(colnames(mse), "alpha")])
mse_interest <- mse[pos_class,]
mse_other <- mse[setdiff(rownames(mse), pos_class),]

library(circlize)
col_fun = colorRamp2(c(-2, 0, 2), hcl_palette = "Blue-Red 3")

ha = HeatmapAnnotation(
    alpha = anno_simple(as.numeric(str_remove(colnames(mse), "alpha "))),
    annotation_name_side = "left")

png("results/rf_mse_interest.png", res = 300, height = 1700, width = 2000)
Heatmap((mse_interest-rowMeans(mse_interest))/matrixStats::rowSds(mse_interest), 
        col = col_fun, show_column_names = F,
         width = ncol(mse_interest)*unit(10, "mm"), 
      height = nrow(mse_interest)*unit(0.2, "mm"),
        cluster_columns = F, show_row_names = F)+ 
  rowAnnotation(class = ifelse(alphas_rf[rownames(mse_interest)]>0, 
                               "class of interest", "no integration"), 
                col = list(class = setNames(c("#70AD47", "grey"), 
                                            nm = c("class of interest", "no integration"))))
dev.off()
#pdf("results/rf_mse_others.pdf", height = 10)
png("results/rf_mse_others.png", res = 300, height = 2200, width = 2000)
Heatmap((mse_other-rowMeans(mse_other))/matrixStats::rowSds(mse_other),
        col = col_fun,show_column_names = F,
        width = ncol(mse_other)*unit(10, "mm"), 
      height = nrow(mse_other)*unit(0.2, "mm"),
        cluster_columns = F, show_row_names = F)+ 
  rowAnnotation(class = ifelse(alphas_rf[rownames(mse_other)]>0, 
                               "class of interest", "no integration"), 
                col = list(class = setNames(c("#70AD47", "grey"), 
                                            nm = c("class of interest", "no integration"))))
dev.off()
```

# For the lasso

```{r, fig.height=10}

load("results/lasso_perumtations_mse_all_genes_test.rdata")
load("results/alpha_per_gene_weighted_LASSO_test_mean_true_shuff_sd.rdata")


pos_class <- names(alphas_lasso[alphas_lasso!=0])

mse <- lmses[str_detect(colnames(lmses), "trueData")]
for(alpha in seq(0,1, by = 0.1)){
  mse[,paste("alpha",alpha)] <- rowMeans(mse[,str_detect(colnames(mse), paste0("LASSO_",as.character(alpha), "_"))])
}
mse <- as.matrix(mse[str_detect(colnames(mse), "alpha")])
mse_interest <- mse[pos_class,]
mse_other <- mse[setdiff(rownames(mse), pos_class),]

library(circlize)
col_fun = colorRamp2(c(-2, 0, 2), hcl_palette = "Blue-Red 3")

ha = HeatmapAnnotation(
    alpha = anno_simple(as.numeric(str_remove(colnames(mse), "alpha "))),
    annotation_name_side = "left")

#pdf("results/lasso_mse_interest.pdf", height = 7)
png("results/lasso_mse_interest.png", res = 300, height = 2000, width = 2000)
Heatmap((mse_interest-rowMeans(mse_interest))/matrixStats::rowSds(mse_interest), 
        col = col_fun, show_column_names = F,
         width = ncol(mse_interest)*unit(10, "mm"), 
      height = nrow(mse_interest)*unit(0.2, "mm"),
        cluster_columns = F, show_row_names = F)+ 
  rowAnnotation(class = ifelse(alphas_lasso[rownames(mse_interest)]>0, 
                               "class of interest", "no integration"), 
                col = list(class = setNames(c("#70AD47", "grey"), 
                                            nm = c("class of interest", "no integration"))))
dev.off()
#pdf("results/lasso_mse_others.pdf", height = 7)
png("results/lasso_mse_others.png", res = 300, height = 2200, width = 2000)
Heatmap((mse_other-rowMeans(mse_other))/matrixStats::rowSds(mse_other),
        col = col_fun,show_column_names = F,
        width = ncol(mse_other)*unit(10, "mm"), 
      height = nrow(mse_other)*unit(0.2, "mm"),
        cluster_columns = F, show_row_names = F)+ 
  rowAnnotation(class = ifelse(alphas_lasso[rownames(mse_other)]>0, 
                               "class of interest", "no integration"), 
                col = list(class = setNames(c("#70AD47", "grey"), 
                                            nm = c("class of interest", "no integration"))))
dev.off()
```


# Intersection between genes with $\alpha > 1$

Are the genes with $\alpha > 1$ (the ones for which we allow TFBM integration) the same between weightedLASSO and weightedRF?

```{r}
venn <- ggVennDiagram(list("weightedRF" = names(alphas_rf[alphas_rf!=0]),
                   "weightedLASSO" = names(alphas_lasso[alphas_lasso!=0])), edge_size = 2)+ 
  scale_color_manual( values = c("#C55A11","#E67F87"))+
  scale_fill_gradient(high="#e6f2ff", low="#e6f2ff")+
  theme(legend.position = "none");venn

# hypergeometric test to assess the significance of the intersect
p_enrich <- phyper(q=351, m = length(names(alphas_rf[alphas_rf!=0])), 
                   n = length(genes) - length(names(alphas_rf[alphas_rf!=0])), 
                   k = length( names(alphas_lasso[alphas_lasso!=0])), lower.tail = F)
p_enrich
ggexport(venn, filename = "results/classes_intersection.pdf", width = 4, height = 4)
```
Not really, even though the intersection is clearly significant.


# Motif enrichment in gene with $\alpha > 1$

Can we find TFBMs significantly enriched in the promoters of genes benefitting from data integration?

## For weightedRF :

```{r}
load("rdata/pwm_prom_jaspar_dap.rdata")

load("results/alpha_per_gene_weighted_LASSO_test_mean_true_shuff_sd.rdata")
load("results/alpha_per_gene_weighted_RF_mean_true_shuff_sd.rdata")

pos_class_rf <- names(alphas_rf[alphas_rf!=0])
pos_class_lasso <- names(alphas_lasso[alphas_lasso!=0])
common <- intersect(pos_class_lasso, pos_class_rf)

known_tfs <- tfs[which(tfs %in% pwm_prom$TF)]

get_number_of_motifs_per_tfs <- function(genes){
  table(pwm_prom[pwm_prom$target %in% genes & 
                   pwm_prom$TF %in% tfs,"TF"])[known_tfs]
}

get_motif_enrichment <- function(gene_list){
  enrichments_per_pwm <- setNames(rep(1, length(known_tfs)), known_tfs)
  n_targets_lasso_in_all <- get_number_of_motifs_per_tfs(genes)
  n_targets_lasso_in_group <- get_number_of_motifs_per_tfs(gene_list)
  n_group_lasso <- length(gene_list)
  
  for(tf in known_tfs){
    
    # number of genes with that motif in all genes
    n_targets_in_all_tf <- n_targets_lasso_in_all[tf]
    
    # number of genes with that motif in the lasso group
    n_targets_lasso_in_group_tf <- n_targets_lasso_in_group[tf]
    p_lasso <- phyper(q=n_targets_lasso_in_group_tf-1,
           m=n_targets_in_all_tf, #white balls
           n=length(genes)-n_targets_in_all_tf, # black balls
           k=n_group_lasso, lower.tail = FALSE)
    
    
    enrichments_per_pwm[tf]<- p_lasso
  }
  enriched_tfs <- names(which(enrichments_per_pwm < 0.05))
  return(DIANE::get_gene_information(enriched_tfs, organism = "Arabidopsis thaliana"))
  
}
get_motif_enrichment(pos_class_rf)
get_motif_enrichment(pos_class_lasso)
get_motif_enrichment(common)
```



# Properties of th gene-specifically optimized GRNs

The following code chunks learns GRNs with the optimal values of alpha determined by our approach, and stores the results.

```{r, eval = F}
nCores = 34
mats <- list()
nrep <- 10
for(rep in 1:nrep){ # exploring inherent variability
  
  mat_lasso <- weightedLASSO_inference(counts, genes, tfs,
                                   alpha = alphas_lasso, N = 50,
                                   tf_expression_permutation = FALSE,
                                   int_pwm_noise = 0, mda_type = "shuffle",
                                   pwm_occurrence = pwm_occurrence,
                                   nCores = nCores)

  mat_lasso_perm <- weightedLASSO_inference(counts, genes, tfs,
                                   alpha = alphas_lasso, N = 50,
                                   tf_expression_permutation = TRUE,
                                   int_pwm_noise = 0, mda_type = "shuffle",
                                   pwm_occurrence = pwm_occurrence,
                                   nCores = nCores)

  mats[[paste0("LASSO_trueData_", rep)]] <- mat_lasso
  mats[[paste0("LASSO_shuffled_", rep)]] <- mat_lasso_perm
  
   mat_rf <- weightedRF_inference(counts, genes, tfs, nTrees = 2000,
                            alpha = alphas_rf,
                            pwm_occurrence = pwm_occurrence,
                            nCores = nCores,
                            importance = "%IncMSE")
    
    mat_rf_perm <- weightedRF_inference(counts, genes, tfs, nTrees = 2000,
                            alpha = alphas_rf, 
                            tf_expression_permutation = TRUE,
                            pwm_occurrence = pwm_occurrence,
                            nCores = nCores,
                            importance = "%IncMSE")
  
  mats[[paste0("RF_trueData_", rep)]] <- mat_rf
  mats[[paste0("RF_shuffled_", rep)]] <- mat_rf_perm
}
save(mats, file = "results/gene_specific_grns_mean_true_shuff_sd.rdata")


edges <- list()
lmses <- data.frame(row.names = genes)
for(name in names(mats)){ 
      for(density in c(0.005,0.01,0.05)){# exploring importance threshold stringency
        
        if(str_detect(name, "LASSO"))
          edges[[paste0(name, '_', density)]] <-
        weightedLASSO_network(mats[[name]], density = density, pwm_occurrence, 
                          genes, tfs, decreasing = TRUE)
        else
          edges[[paste0(name, '_', density)]] <-
        weightedRF_network(mats[[name]], density = density, pwm_occurrence, 
                          genes, tfs)
        lmses[,name] <- mats[[name]]["mse",]
      }
}
save(lmses, file = "results/gene_specific_mse_mean_true_shuff_sd.rdata") 
save(edges, file = "results/gene_specific_edges_mean_true_shuff_sd.rdata") 

# validation against DAP-Seq
settings <- c("model", "dataset", "rep", "density")
val_specific <-
  evaluate_networks(
    edges,
    validation = c("DAPSeq"),
    nCores = 30,
    input_genes = genes,
    input_tfs = tfs
  )
val_specific[, settings] <-
  str_split_fixed(val_specific$network_name, '_', length(settings))
save(val_specific, file = "results/gene_specific_validation_mean_true_shuff_sd.rdata")
```


## Global optimisation of alpha

The following code validates globally inferred GRNs, with a given density, against an experimental gold standard (here, DAP-Seq interactions).


```{r, fig.width=10, fig.height=10}

load(file = "results/100_permutations_bRF_validation_dap_inf.rdata")
val_rf <- val_dap
load(file = "results/100_permutations_lasso_validation_dap.rdata")
val_lasso <- val_dap

precision_lasso <- draw_validation(validation = val_lasso)+
  plot_annotation(title = "weightedLASSO") & 
  theme(plot.title = element_text(size = 20, hjust = 0.5) )

ggexport(precision_lasso, filename = "results/supp_figures/precision_recall_weightedLASSO.pdf", 
         width = 10, height = 10)

precision_rf <- draw_validation(validation = val_rf)+
  plot_annotation(title = "weightedRF") & 
  theme(plot.title = element_text(size = 20, hjust = 0.5) )


ggexport(precision_rf, filename = "results/supp_figures/precision_recall_weightedRF.pdf", 
         width = 10, height = 10)
```


## Gene-specific optimisation of alpha

```{r, warning=TRUE}
settings <- c("model", "dataset", "rep", "density")

load("results/gene_specific_mse_mean_true_shuff_sd.rdata")
load("results/gene_specific_validation_mean_true_shuff_sd.rdata")

val_spec <- val_specific %>%
  separate(network_name, into = settings, sep = "_") %>%
    filter(density == 0.005) %>%
  mutate(density = paste("D =", density),
           model = str_replace(model, "RF", "weightedRF"),
           model = str_replace(model, "LASSO", "weightedLASSO"), 
         alpha_type = "gene-specific")

data_val <- rbind.data.frame(val_rf, val_lasso) %>%
    filter(density %in% c(0.005))%>%
    group_by(model, alpha, dataset, density) %>%
    mutate(mean_precision = mean(precision, na.rm = T),
           sd_precision = sd(precision, na.rm = T),
           mean_recall = mean(recall, na.rm = T),
           sd_recall = sd(recall, na.rm = T),
           density = paste("D =", density),
           model = str_replace(model, "bRF", "weightedRF"),
           model = str_replace(model, "LASSO", "weightedLASSO")) %>%
  dplyr::select(-network_name) %>%
  mutate(alpha = as.numeric(alpha), alpha_type = "global")

pr_curves <- data_val %>%
  ggplot(aes(x=recall, y=precision, 
             color = interaction(dataset, alpha_type),
             fill = interaction(dataset, alpha_type))) +
  geom_point(size = 0.65)+
  geom_ribbon(aes(ymin = mean_precision - sd_precision , 
                    ymax = mean_precision + sd_precision, x=mean_recall  ), alpha = 0.4)+
  theme_pubr()+ ggh4x::facet_nested_wrap(vars(model), nest_line = T)+
  geom_line(aes(y=mean_precision, x=mean_recall), size=2)+
  theme(strip.background = element_blank())+
  geom_hline(color = "#C55A11", yintercept = 0.331042, size= 1.5, show.legend = T) +
  geom_point(aes(x=recall, y=precision), data = val_spec)+
  scale_color_manual(name = "Dataset and type of integration",
                     values = setNames(c("lightblue", "grey", "#4670CD", "#70AD47"), 
                                       c("shuffled.gene-specific","shuffled.global",
                                         "trueData.gene-specific", "trueData.global")))+
  scale_fill_manual(name = "Dataset and type of integration",
                     values = setNames(c("lightblue", "grey", "#4670CD", "#70AD47"), 
                                       c("shuffled.gene-specific","shuffled.global",
                                         "trueData.gene-specific", "trueData.global")));pr_curves
```
Figure including permutations :


```{r, fig.width=12, fig.height=10}

plot_MSE_gene_specific <- function(model_){
  lmses %>%
  rownames_to_column("gene") %>%
  reshape2::melt()%>%
    separate(variable,
             into = c("model", "dataset",  "rep"),
             sep = "_") %>%
  filter(model == model_) %>%
  group_by(dataset, rep) %>%
  summarise(median_MSE = median(value, na.rm=T))%>%
    ggplot(aes(x=dataset, y = median_MSE, color = dataset, fill = dataset)) + 
  scale_color_manual(values = setNames(c("lightblue", "#4670CD"), c("shuffled", "trueData")))+
    scale_fill_manual(values = setNames(c("lightblue", "#4670CD"), c("shuffled", "trueData"))) +
  geom_boxplot(size = 1, alpha = 0.5, outlier.alpha = 0) +
  theme_pubr() + geom_jitter(width = 0.2, size = 2)+
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    axis.ticks.x = element_blank(),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text.x = element_blank(),
    legend.position = 'none'
  )+ xlab("")+ ylab("MSE")
}


# gene specific MSE
mse_lasso_spec <- plot_MSE_gene_specific("LASSO")
mse_rf_spec <- plot_MSE_gene_specific("RF")


# MSE for global alphas
draw_mse <- function(model){
data <- lmses[as.numeric(str_split_fixed(colnames(lmses), '_', 4)[,4]) <=10] %>%
rownames_to_column("gene") %>%
reshape2::melt()%>%
  separate(variable,
           into = c("model", "alpha", "dataset",  "rep"),
           sep = "_") %>%
filter(model == model) %>%
group_by(dataset, rep, alpha) %>%
summarise(median_MSE = median(value, na.rm=T))%>%
  group_by(alpha, dataset) %>%
  mutate(mean_median_mse = mean(median_MSE),
         sd_median_mse = sd(median_MSE)) %>%
  mutate(alpha = as.numeric(alpha))
data %>%
  ggplot(aes(x=alpha, y = median_MSE, color = dataset, fill = dataset)) + 
scale_color_manual(values = setNames(c("grey", "#70AD47"),
                                     c("shuffled", "trueData")))+
  scale_fill_manual(values = setNames(c("grey", "#70AD47"), 
                                      c("shuffled", "trueData"))) +
  geom_line(aes(y=mean_median_mse, group = dataset)) +
  geom_ribbon(aes(ymin = mean_median_mse - sd_median_mse , 
                    ymax = mean_median_mse + sd_median_mse), 
                alpha = .4)+ xlab(expression(alpha)) +
theme_pubr() + geom_point(width = 0.2, size = 2)+
theme(
  strip.background = element_blank(),
  axis.title.x = element_text(size = 22),
  title = element_text(size = 16),
  strip.text = element_text(size = 16),
  legend.text = element_text(size = 15),
  legend.position = 'none'
)+ ylab("MSE")
}

load("results/lasso_perumtations_mse_all_genes_test.rdata")
mse_lasso <- draw_mse("LASSO")

load("results/brf_perumtations_mse_all_genes_predict.rdata")
mse_rf <- draw_mse("RF")
x_min <- 0.19

plot <- (mse_lasso + ylim(c(x_min,0.265)) +labs(subtitle = "Global\ndata integration")+ 
  mse_lasso_spec+ ylim(c(x_min,0.265)) +labs(subtitle = "Gene-specific\ndata integration")+ylab("") +
  mse_rf +  ylim(c(x_min,0.265)) +labs(subtitle = "Global\ndata integration")+ ylab("") +
  mse_rf_spec+ ylim(c(x_min,0.265))  +labs(subtitle = "Gene-specific\ndata integration")+ylab(""))+
  plot_layout(guides = "collect", ncol = 4, widths = c(1.5,1,1.5,1)) & theme(legend.position = 'bottom')

ggexport(plot, filename = "results/specific_grns_mse_with_baseline.pdf", width = 11, height = 6)

plot_final <- plot/ pr_curves+theme(legend.position = "none", 
                      axis.title = element_text(size= 15),
                      strip.text = element_text(size = 18))+
  plot_layout(heights = c(1,1.5))

ggexport(plot_final, filename = "results/specific_grns_mean_true_shuff_sd.pdf", width = 11, height = 10)

```


Figure without permutations :


```{r, fig.width=12, fig.height=10}

plot_MSE_gene_specific <- function(model_){
  lmses %>%
rownames_to_column("gene") %>%
reshape2::melt()%>%
  separate(variable,
           into = c("model", "dataset",  "rep"),
           sep = "_") %>%
filter(model == model_ & dataset == "trueData") %>%
group_by(dataset , rep) %>%
summarise(median_MSE = median(value, na.rm=T))%>%
  ggplot(aes(x=dataset, y = median_MSE)) + 
geom_boxplot(size = 1, alpha = 0.5, outlier.alpha = 0, 
             color = "#4670CD", fill = "#4670CD") +
theme_pubr() + geom_jitter(width = 0.2, size = 2,  color = "#4670CD")+
theme(
  strip.background = element_blank(),
  axis.line = element_blank(), 
  axis.title=element_blank(),
  axis.text=element_blank(),
  axis.ticks=element_blank(), 
  title = element_text(size = 16),
  strip.text = element_text(size = 16),
  legend.text = element_text(size = 15),
  legend.position = 'none'
)+ xlab("")+ ylab("MSE")
}
# gene specific MSE
mse_lasso_spec <- plot_MSE_gene_specific("LASSO")
mse_rf_spec <- plot_MSE_gene_specific("RF")

# MSE for global alphas
draw_mse <- function(model){
data <- lmses[as.numeric(str_split_fixed(colnames(lmses), '_', 4)[,4]) <=10] %>%
rownames_to_column("gene") %>%
reshape2::melt()%>%
  separate(variable,
           into = c("model", "alpha", "dataset",  "rep"),
           sep = "_") %>%
filter(model == model & dataset == "trueData") %>%
group_by(dataset, rep, alpha) %>%
summarise(median_MSE = median(value, na.rm=T))%>%
  group_by(alpha, dataset) %>%
  mutate(mean_median_mse = mean(median_MSE),
         sd_median_mse = sd(median_MSE)) %>%
  mutate(alpha = as.numeric(alpha))
data %>%
  ggplot(aes(x=alpha, y = median_MSE)) + 
  geom_line(aes(y=mean_median_mse, group = dataset), color = "#70AD47") +
  geom_ribbon(aes(ymin = mean_median_mse - sd_median_mse , 
                    ymax = mean_median_mse + sd_median_mse), 
                alpha = .4, color = "#70AD47", fill = "#70AD47")+ 
  xlab(expression(alpha)) +
theme_pubr() + geom_point(width = 0.2, size = 2, color = "#70AD47")+
theme(
  strip.background = element_blank(),
  axis.title.x = element_text(size = 22),
  title = element_text(size = 16),
  strip.text = element_text(size = 16),
  legend.text = element_text(size = 15),
  legend.position = 'top'
)+ ylab("MSE")
}

load("results/lasso_perumtations_mse_all_genes_test.rdata")
mse_lasso <- draw_mse("LASSO")

load("results/brf_perumtations_mse_all_genes_predict.rdata")
mse_rf <- draw_mse("RF")
x_min <- 0.19


plot <- (mse_lasso + ylim(c(x_min,0.26)) +labs(title = "weightedLASSO")+ 
           theme(plot.title = element_text(hjust = 1))+
  mse_lasso_spec+ ylim(c(x_min,0.26)) +ylab("") +
  mse_rf +  ylim(c(x_min,0.26)) +labs(title = "weightedRF")+
    theme(plot.title = element_text(hjust = 0.75))+ylab("") +
  mse_rf_spec+ ylim(c(x_min,0.26)) + ylab(""))+
  plot_layout(guides = "collect", ncol = 4, widths = c(3,1,3,1)) & 
  theme(legend.position = 'bottom', legend.text = element_text(size = 15))


pr_curves <- data_val %>%
  filter(dataset == "trueData") %>%
  ggplot(aes(x=recall, y=precision, 
             label = alpha, fill = alpha_type, color = alpha_type)) +
  geom_point(size = 0.65)+
  geom_ribbon(aes(ymin = mean_precision - sd_precision , 
                    ymax = mean_precision + sd_precision, x=mean_recall  ), alpha = 0.4)+
  theme_pubr()+ ggh4x::facet_nested_wrap(vars(model), nest_line = T)+
  geom_line(aes(y=mean_precision, x=mean_recall), size=2)+
  geom_label(aes(y=mean_precision, x=mean_recall), nudge_y = 0.02, fill = "white", show.legend = F)+
  theme(strip.background = element_blank(), 
        legend.text = element_text(size = 15),
        legend.title = element_text(size = 18))+
  geom_hline(color = "#C55A11", linetype='dashed',
             yintercept = 0.331042, size= 1.5, show.legend = T) +
  geom_point(aes(x=recall, y=precision), data = val_spec[val_spec$dataset=="trueData",])+
  scale_color_manual(name = "Data integration",
                     values = setNames(c( "#4670CD", "#70AD47"), 
                                       c("gene-specific", "global")))+
  scale_fill_manual(name = "Data integration",
                     values = setNames(c( "#4670CD", "#70AD47"), 
                                       c("gene-specific", "global")))+
  xlab("Recall") + ylab("Precision")


plot_final <- plot/ pr_curves+theme(legend.position = "bottom", 
                      axis.title = element_text(size= 15),
                      strip.text = element_text(size = 18))+
  plot_layout(heights = c(1,1));plot_final


plot_final

ggexport(plot_final, filename = "results/specific_grns_no_permutations_mean_sds.pdf", 
         width = 11, height = 10)

```


# Comparison of our approach with the simplest min MSE criterion


Intersection between genes benefiting from data integration with the minimal MSE criterion.

```{r}

load("results/alpha_per_gene_weighted_LASSO_min.rdata")
load("results/alpha_per_gene_weighted_RF_min.rdata")

venn <- ggVennDiagram(list("weightedRF" = names(alphas_rf[alphas_rf!=0]),
                   "weightedLASSO" = names(alphas_lasso[alphas_lasso!=0])), edge_size = 2)+ 
  scale_color_manual( values = c("#C55A11","#E67F87"))+
  scale_fill_gradient(high="#e6f2ff", low="#e6f2ff")+
  theme(legend.position = "none");venn

# hypergeometric test to assess the significance of the intersect
p_enrich <- phyper(q=582, m = length(names(alphas_rf[alphas_rf!=0])), 
                   n = length(genes) - length(names(alphas_rf[alphas_rf!=0])), 
                   k = length( names(alphas_lasso[alphas_lasso!=0])), lower.tail = F)
p_enrich
ggexport(venn, filename = "results/classes_intersection_min.pdf", width = 4, height = 4)
```

## MSE, precision and recall plots

don't know if it will be used

```{r, fig.width=10, fig.height=10, eval = F}
load("results/gene_specific_grns_min.rdata")
load("results/gene_specific_mse_min.rdata")
load("results/gene_specific_edges_min.rdata")

load("results/gene_specific_validation_min.rdata")

spec_rf <- val_specific %>%
  separate(network_name, into = settings, sep = "_") %>%
    filter(density == 0.005 & model == "RF") %>%
  mutate(density = paste("D =", density),
           model = str_replace(model, "RF", "weightedRF"),
           model = str_replace(model, "LASSO", "weightedLASSO"))%>%
  ggplot(aes(x=dataset, y = precision, color = dataset, fill = dataset)) + 
  scale_color_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData")))+
    scale_fill_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData"))) +
  geom_hline(color = "#C55A11", yintercept = 0.331042, size= 1.5, show.legend = T) + 
  geom_boxplot(size = 1, alpha = 0.5, outlier.alpha = 0) +
  theme_pubr() + geom_jitter(width = 0.2, size = 2)+
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_blank(),
    legend.position = 'none'
  ) + xlab("") + ylim(c(0.2,0.46))+ ylab("")

spec_lasso <- val_specific %>%
  separate(network_name, into = settings, sep = "_") %>%
    filter(density == 0.005 & model == "LASSO") %>%
  mutate(density = paste("D =", density),
           model = str_replace(model, "RF", "weightedRF"),
           model = str_replace(model, "LASSO", "weightedLASSO"))%>%
  ggplot(aes(x=dataset, y = precision, color = dataset, fill = dataset)) + 
  scale_color_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData")))+
    scale_fill_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData"))) +
  geom_hline(color = "#C55A11", yintercept = 0.331042, size= 1.5, show.legend = T) + 
  geom_boxplot(size = 1, alpha = 0.5, outlier.alpha = 0) +
  theme_pubr() + geom_jitter(width = 0.2, size = 2)+
  theme(
    strip.background = element_blank(),
    axis.title.x = element_text(size = 22),
    title = element_text(size = 16),
    strip.text = element_text(size = 16),
    legend.text = element_text(size = 15),
    axis.text = element_blank(),
    legend.position = 'none'
  )+ xlab("")+ ylim(c(0.2,0.46))+ ylab("")


mse_lasso_spec <- lmses %>%
rownames_to_column("gene") %>%
reshape2::melt()%>%
  separate(variable,
           into = c("model", "dataset",  "rep"),
           sep = "_") %>%
filter(model == "LASSO") %>%
group_by(dataset, rep) %>%
summarise(median_MSE = median(value, na.rm=T))%>%
  ggplot(aes(x=dataset, y = median_MSE, color = dataset, fill = dataset)) + 
scale_color_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData")))+
  scale_fill_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData"))) +
geom_boxplot(size = 1, alpha = 0.5, outlier.alpha = 0) +
theme_pubr() + geom_jitter(width = 0.2, size = 2)+
theme(
  strip.background = element_blank(),
  axis.title.x = element_text(size = 22),
  title = element_text(size = 16),
  strip.text = element_text(size = 16),
  legend.text = element_text(size = 15),
  axis.text.x = element_blank(),
  legend.position = 'none'
)+ xlab("")+ ylab("")

mse_rf_spec <- lmses %>%
rownames_to_column("gene") %>%
reshape2::melt()%>%
  separate(variable,
           into = c("model", "dataset",  "rep"),
           sep = "_") %>%
filter(model == "RF") %>%
group_by(dataset, rep) %>%
summarise(median_MSE = median(value, na.rm=T))%>%
  ggplot(aes(x=dataset, y = median_MSE, color = dataset, fill = dataset)) + 
scale_color_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData")))+
  scale_fill_manual(values = setNames(c("grey", "#70AD47"), c("shuffled", "trueData"))) +
geom_boxplot(size = 1, alpha = 0.5, outlier.alpha = 0) +
theme_pubr() + geom_jitter(width = 0.2, size = 2)+
theme(
  strip.background = element_blank(),
  axis.title.x = element_text(size = 22),
  title = element_text(size = 16),
  strip.text = element_text(size = 16),
  legend.text = element_text(size = 15),
  axis.text.x = element_blank(),
  legend.position = 'none'
)+ xlab("")+ ylab("") 


mse_lasso + ylim(c(0.17,0.265)) +labs(subtitle = "LASSO")+ ggtitle("MSE") + 
  mse_lasso_spec+ ylim(c(0.17,0.265)) +labs(subtitle = "LASSO")+
  mse_rf +  ylim(c(0.17,0.265)) +labs(subtitle = "RF") +
  mse_rf_spec+ ylim(c(0.17,0.265))  +labs(subtitle = "RF")+
  precision_lasso + ylim(c(0.2,0.46)) + ggtitle("Precision") + labs(subtitle = "LASSO")+
  theme(legend.position = 'none') + 
  spec_lasso + labs(subtitle = "LASSO")+
  precision_rf + labs(subtitle = "RF") +ylim(c(0.2,0.46))+
  theme(legend.position = 'none')+ggtitle("") + 
  spec_rf +labs(subtitle = "RF") +
plot_layout(guides = "collect", ncol = 4, widths = c(1,1,1,1)) 
```

# Do genes lost by our metric have a lower precision?


## Differences between gene lists for the two criteria : Maximal divergence from permutations and minimal MSE.

Differences between the two criteria for the same model : 

```{r}
alphas_lasso_min <- alphas_lasso
alphas_rf_min <- alphas_rf

load("results/alpha_per_gene_weighted_LASSO_test.rdata")
load("results/alpha_per_gene_weighted_RF.rdata")

alphas_lasso_div <- alphas_lasso
alphas_rf_div <- alphas_rf

saved_lasso <- setdiff(names(alphas_lasso_div[alphas_lasso_div > 0]), 
                             names(alphas_lasso_min[alphas_lasso_min > 0]))

saved_rf <- setdiff(names(alphas_rf_div[alphas_rf_div > 0]), 
                             names(alphas_rf_min[alphas_rf_min > 0]))

same_rf <- intersect(names(alphas_rf_div[alphas_rf_div > 0]), 
                             names(alphas_rf_min[alphas_rf_min > 0]))

lost_lasso <- setdiff(names(alphas_lasso_min[alphas_lasso_min > 0]),
                            names(alphas_lasso_div[alphas_lasso_div > 0]))

lost_rf <- setdiff(names(alphas_rf_min[alphas_rf_min > 0]),
                         names(alphas_rf_div[alphas_rf_div > 0]))

same_lasso <- intersect(names(alphas_lasso_min[alphas_lasso_min > 0]),
                        names(alphas_lasso_div[alphas_lasso_div > 0]))

matrix(c(length(same_lasso), length(lost_lasso), length(saved_lasso), 
         length(genes) - length(same_lasso) - length(lost_lasso) - length(saved_lasso)), 
       nrow = 2, byrow = F, dimnames = list(c("data integration div", "no data integration div"),
                                            c("data integration min", "no data integration min")))


```

For the RFs :

```{r}
matrix(c(length(same_rf), length(lost_rf), length(saved_rf), 
         length(genes) - length(same_rf) - length(lost_rf) - length(saved_rf)), 
       nrow = 2, byrow = F, dimnames = list(c("data integration div", "no data integration div"),
                                            c("data integration min", "no data integration min")))
```

# Comparing the precision and recall of min and dev GRNs

```{r, fig.width=10}
load("results/gene_specific_validation_min.rdata")
val_specific_min <- val_specific %>%
  mutate(criterion = "min MSE")
load("results/gene_specific_validation_mean_true_shuff_sd.rdata")
val_specific_dev <- val_specific%>%
  mutate(criterion = "proposed approach")

rbind.data.frame(val_specific_dev, val_specific_min) %>%
  filter(dataset == "trueData" & density < 0.05) %>%
  ggplot(aes(y = precision, x = recall, color = criterion, label = density)) + 
  geom_jitter()+
  facet_nested_wrap(vars(model, density), ncol = 2, scales = "free") + theme_bw()+
  theme(strip.background = element_blank()) +
  stat_ellipse()
```

By using our approach over the min MSE approach to chose alpha, we lose some edges and gain some. Do we loose low precision/recall edges and gain high precision/recall edges?


```{r, eval=T, fig.width=10}

load("results/gene_specific_edges_min.rdata")
edges_min <- edges

load(file = "results/gene_specific_edges_mean_true_shuff_sd.rdata")
edges_dev <- edges


edges_diff <- list()
for(name in names(edges)){
  print(name)
  edges_diff[[paste0(name, "_gained")]] <- setdiff(edges_dev[[name]], edges_min[[name]])
  edges_diff[[paste0(name, "_lost")]] <- setdiff(edges_min[[name]], edges_dev[[name]])
}

settings = c("model", "dataset", "rep", "density", "edge")

val_diff <- evaluate_networks(
    edges_diff,
    validation = c("DAPSeq"),
    nCores = 30,
    input_genes = genes,
    input_tfs = tfs)

val_diff[, settings] <-
  str_split_fixed(val_diff$network_name, '_', length(settings))

val_diff %>%
  filter(dataset == "trueData" & 
           density < 0.05) %>%
  ggplot(aes(x = recall, y = precision, color = edge)) + 
  geom_point()+
  facet_nested_wrap(vars(model, density), ncol = 2, scales = "free") + theme_bw()+
  theme(strip.background = element_blank()) + stat_ellipse()
```

Shows the top 25 regulators in terms of out-degree.

```{r}
load("results/gene_specific_edges_mean_true_shuff_sd.rdata")

annot <- DIANE::get_gene_information(genes,organism = "Arabidopsis thaliana")

get_hubs <- function(nets, metric = "degree"){
  g <- setNames(rep(0, length(genes)), genes)
  for(net in names(nets)){
    if(metric == "degree")
      g[names(table(nets[[net]]$from))] <- g[names(table(nets[[net]]$from))] + table(nets[[net]]$from)/length(nets)
    if(metric == "betweenness")
      g[names(table(nets[[net]]$from))]<- g[names(table(nets[[net]]$from))]+
        igraph::betweenness(igraph::graph_from_data_frame(nets[[net]], directed = T),
                            v = names(table(nets[[net]]$from)), directed = T)
  }
  df <- data.frame(genes = genes, average_degree = g[genes], 
                   label = annot[match(genes, rownames(annot)), "label"],
                   pwm_available =  genes %in% known_tfs)
  
  df$regulator <- ifelse(df$label %in% c("DIV1", "TGA1", "TGA4", "HHO2", "HHO3", "HRS1", "BT1", "BT2"), 
                         "Nitrate regulator",
                         ifelse(df$label %in%  c("VRN1", "CRF4"), "Candidate nitrate\nregulator",
                                ifelse(df$label %in% c("NLP7", "PHL1", "BZIP53", "HHO6", "JKD"),
                                       "Retreived by gene\nspecific data integration\n optimisation", "No knowlege")))
  
  
  plt <- df %>%
    slice_max(average_degree, n = 25) %>% 
    ggplot(aes(x=average_degree, label = label, color = regulator,
               y=reorder(label, average_degree)))+ 
    geom_segment(x=0, color = "grey",
                 aes(xend=average_degree, 
                          yend = reorder(label, average_degree)))+
    geom_point()+
    geom_text(nudge_x = 7, show.legend = F)+ theme_bw() + 
    ylab("") + xlab("Average out-degree")+theme(axis.text.y = element_blank())+
    scale_color_manual(name = "", values = setNames(c("#C55A11", "orange", "#4670CD", "black"), 
                                         c("Nitrate regulator", "Candidate nitrate\nregulator",
                                           "Retreived by gene\nspecific data integration\n optimisation", "No knowledge")))+
    guides(color = guide_legend(nrow=4,byrow=TRUE))
  
  top20<- df %>%
    slice_max(average_degree, n = 25) %>%
    select(label)
  
  return(list(ranking = plt, top25 = top20$label))
}

rf_nets <- names(edges)[str_detect(names(edges),"RF_trueData") ]
rf_nets <- edges[rf_nets[str_detect(rf_nets,"0.005") ]]


lasso_nets <- names(edges)[str_detect(names(edges),"LASSO_trueData") ]
lasso_nets <- edges[lasso_nets[str_detect(lasso_nets,"0.005") ]]


```
Building global networks :

```{r}

load(file = "results/100_permutations_lasso_edges.rdata")

lasso_nets_0 <- names(edges)[str_detect(names(edges),"LASSO_0_trueData") ]
lasso_nets_0 <- edges[lasso_nets_0[str_detect(lasso_nets_0,"0.005") ]]

lasso_nets_1 <- names(edges)[str_detect(names(edges),"LASSO_1_trueData") ]
lasso_nets_1 <- edges[lasso_nets_1[str_detect(lasso_nets_1,"0.005") ]]


load(file = "results/100_permutations_bRF_edges_inf.rdata")

rf_nets_0 <- names(edges)[str_detect(names(edges),"RF_0_trueData") ]
rf_nets_0 <- edges[rf_nets_0[str_detect(rf_nets_0,"0.005") ]]

rf_nets_1 <- names(edges)[str_detect(names(edges),"RF_1_trueData") ]
rf_nets_1 <- edges[rf_nets_1[str_detect(rf_nets_1,"0.005") ]]

```


Rankings in global alpha networks compared to alpha specific in RF:

```{r}

ranks_rf <- get_hubs(rf_nets_0)$ranking+ ggtitle("weightedRF alpha = 0") + 
  theme(legend.position = "none")+
  get_hubs(rf_nets_1)$ranking+ggtitle("weightedRF alpha = 1") + 
  theme(legend.position = "none")+
  get_hubs(rf_nets)$ranking+ ggtitle("weightedRF specific alpha")+
  theme(legend.position = "top")+
  plot_layout(guides = "collect") 



ranks_rf <- get_hubs(rf_nets)$ranking+ ggtitle("weightedRF, gene-specific optimsation")+
  theme(legend.position = "right", legend.text = element_text(size = 15))

ggexport(ranks_rf, filename = "results/weightedRF_ranks_grns.pdf", width = 8, ehight = 7)
# regulators in the top ranking of all GRNs
common_rf <- intersect(get_hubs(rf_nets)$top25 , union(get_hubs(rf_nets_0)$top25,
                                        get_hubs(rf_nets_1)$top25))

# regulators in gene-specific GRNs only
setdiff(get_hubs(rf_nets)$top25 , union(get_hubs(rf_nets_0)$top25,
                                        get_hubs(rf_nets_1)$top25))

```

Rankings in global alpha networks compared for alpha specific in LASSO:

```{r}


get_hubs(lasso_nets_0)$ranking+ ggtitle("weightedLASSO alpha = 0") +
  theme(legend.position = "none")+
  get_hubs(lasso_nets_1)$ranking+ggtitle("weightedLASSO alpha = 1") +
  theme(legend.position = "none")+
  get_hubs(lasso_nets)$ranking+ ggtitle("weightedLASSO specific alpha")+
  theme(legend.position = "top")+
  plot_layout(guides = "collect") 

# regulators in the top ranking of all GRNs
common_lasso <- intersect(get_hubs(lasso_nets)$top25 , union(get_hubs(lasso_nets_0)$top25,
                                        get_hubs(lasso_nets_1)$top25))

# regulators that were not in the top 25
setdiff(get_hubs(lasso_nets)$top25 , 
        union(get_hubs(lasso_nets_0)$top25,
              get_hubs(lasso_nets_1)$top25))


```

```{r}
intersect(common_lasso, common_rf)
```


```{r}

get_hubs(rf_nets_0, metric = "betweenness")$ranking+ ggtitle("weightedRF alpha = 0") + 
  get_hubs(rf_nets_1, metric = "betweenness")$ranking+ggtitle("weightedRF alpha = 1") + 
  get_hubs(rf_nets, metric = "betweenness")$ranking+ ggtitle("weightedRF specific alpha")

# intersect(get_hubs(rf_nets)$top25 , get_hubs(rf_nets_1)$top25)
# intersect(get_hubs(rf_nets)$top25 , get_hubs(rf_nets_0)$top25)

intersect(get_hubs(rf_nets, metric = "betweenness")$top25 , union(get_hubs(rf_nets_0, metric = "betweenness")$top25,
                                        get_hubs(rf_nets_1, metric = "betweenness")$top25))

# regulators that were not in the top 25
setdiff(get_hubs(rf_nets, metric = "betweenness")$top25 , union(get_hubs(rf_nets_0, metric = "betweenness")$top25,
                                        get_hubs(rf_nets_1, metric = "betweenness")$top25))


```







Getting the most regulated genes in specific GRNs :


```{r}
get_hubs_incoming <- function(nets){
  g <- setNames(rep(0, length(genes)), genes)
  for(net in names(nets)){
    g[names(table(nets[[net]]$to))] <- g[names(table(nets[[net]]$to))] + table(nets[[net]]$to)/length(nets)
  }
  df <- data.frame(genes = genes, average_degree = g[genes], 
                   label = annot[match(genes, rownames(annot)), "label"])
  plt <- df %>%
    slice_max(average_degree, n = 25) %>%
    ggplot(aes(x=average_degree, y=reorder(label, average_degree) ))+ 
    geom_point() + theme_bw() + ylab("gene") + xlab("Average in-degree")
  
  top20<- df %>%
    slice_max(average_degree, n = 25) %>%
    select(label)
  return(list(ranking = plt, top20 = top20$label))
}

get_hubs_incoming(rf_nets)$ranking+ ggtitle("weightedRF") + 
  get_hubs_incoming(lasso_nets)$ranking+ ggtitle("weightedLASSO")
intersect(get_hubs_incoming(rf_nets)$top25 , get_hubs_incoming(lasso_nets)$top25)
```

Clustering coefficient:




```{r, eval = T}

load("results/gene_specific_edges_mean_true_shuff_sd.rdata")


rf_nets <- names(edges)[str_detect(names(edges),"RF_trueData") ]
rf_nets <- edges[rf_nets[str_detect(rf_nets,"0.005") ]]

lasso_nets <- names(edges)[str_detect(names(edges),"LASSO_trueData") ]
lasso_nets <- edges[lasso_nets[str_detect(lasso_nets,"0.005") ]]


get_transitivity <- function(nets, settings){
  res <- data.frame(trans = unlist(lapply(names(nets), 
           function(net){return(igraph::transitivity(igraph::graph_from_data_frame(nets[[net]], directed = T), 
                                                     type = "global"))})),
           length = unlist(lapply(names(nets), 
           function(net){return(igraph::mean_distance(igraph::graph_from_data_frame(nets[[net]], directed = T)))})),
           diam = unlist(lapply(names(nets), 
           function(net){return(igraph::diameter(igraph::graph_from_data_frame(nets[[net]], directed = T)))})))
  
  res$settings <- settings
  return(res)
}

trans <- rbind.data.frame(get_transitivity(rf_nets_0, "weightedRF_0"), 
                 get_transitivity(rf_nets_1, "weightedRF_1"), 
                 get_transitivity(rf_nets, "weightedRF_specific"),
                 get_transitivity(lasso_nets_0, "weightedLASSO_0"), 
                 get_transitivity(lasso_nets_1, "weightedLASSO_1"), 
                 get_transitivity(lasso_nets, "weightedLASSO_specific"))

trans <- trans %>%
  separate(settings, into = c("model", "alpha"), sep = "_")

trans %>%
  ggplot(aes(x=alpha, y = trans, color = alpha)) + 
  ggh4x::facet_nested_wrap(vars(model), nrow = 2, scales = "free")+ 
  theme_bw()+ 
  theme(strip.background = element_blank())+
  geom_point() +ggtitle("clustering coeff")

trans %>%
  ggplot(aes(x=alpha, y = length, color = alpha)) + 
  ggh4x::facet_nested_wrap(vars(model), nrow = 2, scales = "free")+ 
  theme_bw()+ 
  theme(strip.background = element_blank())+
  geom_point() +ggtitle("avg path length coeff")

trans %>%
  ggplot(aes(x=alpha, y = diam, color = alpha)) + 
  ggh4x::facet_nested_wrap(vars(model), nrow = 2, scales = "free")+ 
  theme_bw()+ 
  theme(strip.background = element_blank())+
  geom_point() +ggtitle("diameter")

```


Nitrate GSEA :

```{r, eval = T}


N_go <- c("GO:0006807", "GO:0051171", "GO:0051173", "GO:0022622", "GO:0034641", "GO:0044271")

convert_from_agi <- function(ids){
  x <- org.At.tair.db::org.At.tairENTREZID
  mapped_genes <- AnnotationDbi::mappedkeys(x)
  xx <- as.list(x[mapped_genes])
  return(unlist(xx[as.vector(ids)]))
}
# 
# load("results/gene_specific_edges_mean_true_shuff_sd.rdata")
# 
# 
# mean(edges$LASSO_trueData_1_0.005$pwm)
# mean(lasso_nets_0$LASSO_0_trueData_1_0.005$pwm)
# mean(lasso_nets_1$LASSO_1_trueData_1_0.005$pwm)
# 
# mean(edges$RF_trueData_10_0.005$pwm)
# mean(rf_nets_0$bRF_0_trueData_1_0.005$pwm)
# mean(rf_nets_1$bRF_1_trueData_1_0.005$pwm)
# 
# 
# rf_nets <- names(edges)[str_detect(names(edges),"RF_trueData") ]
# rf_nets <- edges[rf_nets[str_detect(rf_nets,"0.005") ]]


# lasso_nets <- names(edges)[str_detect(names(edges),"LASSO_trueData") ]
# lasso_nets <- edges[lasso_nets[str_detect(lasso_nets,"0.005") ]]


get_gsea <- function(nets, settings){
  g <- setNames(rep(0, length(genes)), genes)
  for(net in names(nets)){
    
    # g[names(table(nets[[net]]$to))]<-g[names(table(nets[[net]]$to))]+
    #   igraph::betweenness(igraph::graph_from_data_frame(nets[[net]], directed = T),
    #                       v = names(table(nets[[net]]$to)), directed = F)
    
    # g[names(table(nets[[net]]$to))]<-g[names(table(nets[[net]]$to))]+
    #   igraph::transitivity(igraph::graph_from_data_frame(nets[[net]], directed = T),
    #                       vids = names(table(nets[[net]]$to)), type = "local")
    #g[names(table(nets[[net]]$from))] <- g[names(table(nets[[net]]$from))] + table(nets[[net]]$from)/length(nets)
    
    # g[genes] <- g[genes] + (table(nets[[net]]$from)[genes]+table(nets[[net]]$to)[genes])/length(nets)
    ge <- intersect(genes, unique(c(nets[[net]]$from, nets[[net]]$to)))
    g[ge]<-g[ge]+
      igraph::degree(igraph::graph_from_data_frame(nets[[net]], directed = T),
                          v = ge, mode = "total")
  }
  df <- data.frame(genes = genes, average_degree = g[genes],
                   label = annot[match(genes, rownames(annot)), "label"])


  # degrees <- sort(setNames(df$average_degree,convert_from_agi(df$average_degree)),
  #                 decreasing = T)
  df <- df[df$average_degree>0,]
  degrees <- sort(setNames(df$average_degree,convert_from_agi(df$genes)),
                  decreasing = T)
  
  degree <- degrees[!is.na(names(degrees))]
  
  gse <- gseGO(geneList=degrees,
              ont ="BP",
              maxGSSize = length(degrees)-1, nPerm =1000,
              pvalueCutoff = 1,
              verbose = TRUE, scoreType = "pos",
              OrgDb = org.At.tair.db::org.At.tair.db,
              pAdjustMethod = "BH")
  
  res <- gse@result[gse@result$ID %in% N_go,c("Description", "enrichmentScore", "p.adjust")]
  res$settings = settings
  return(res)
}

gsea_nitrogen <- rbind.data.frame(get_gsea(rf_nets_0, "weightedRF_0"), 
                 get_gsea(rf_nets_1, "weightedRF_1"), 
                 get_gsea(rf_nets, "weightedRF_specific"),
                 get_gsea(lasso_nets_0, "weightedLASSO_0"), 
                 get_gsea(lasso_nets_1, "weightedLASSO_1"), 
                 get_gsea(lasso_nets, "weightedLASSO_specific"))

gsea_nitrogen <- gsea_nitrogen %>%
  separate(settings, into = c("model", "alpha"), sep = "_")
# res_rf <- res
# 
# gs <- rbind.data.frame(res_rf_0, res_rf_1, res_rf)
# gs$alpha<- c(rep("alpha_0",4), rep("alpha_1", 4), rep("alpha_specific",4))
# gs$model <- "weightedRF"
# 
# gs_lasso <- rbind.data.frame(res_lasso_0, res_lasso_1, res_lasso)
# gs_lasso$alpha <- c(rep("alpha_0",4), rep("alpha_1", 4), rep("alpha_specific",4))
# gs_lasso$model <- "weightedLASSO"
# 

# gsea_nitrogen <- rbind.data.frame(gs, gs_lasso) 
# save(gsea_nitrogen, file = "results/gsea_nitrogen.rdata")


gsea_nitrogen %>%
  ggplot(aes(x=alpha, y = enrichmentScore, color = Description)) + 
  ggh4x::facet_nested_wrap(vars(Description,model), nrow = 2)+ theme_bw()+ 
  theme(strip.background = element_blank())+
  geom_point() +ggtitle("Enrichment score (GSEA)")


gsea_nitrogen %>%
  ggplot(aes(x=alpha, y = -log(p.adjust), color = Description)) + 
  ggh4x::facet_nested_wrap(vars(Description,model), nrow = 2)+ theme_bw()+ 
  theme(strip.background = element_blank())+
  geom_point() + ggtitle("Enrichment p adjusted pavlue (GSEA)")

```
# degree distributions of TFs with and without PWMs

```{r}

# load("rdata/pwm_prom_jaspar_dap.rdata")
known_tfs <- intersect(genes,unique(pwm_prom$TF))
get_degree_hist <- function(nets, title){
  g <- setNames(rep(0, length(genes)), genes)
  for(net in names(nets)){
    g[names(table(nets[[net]]$to))] <- g[names(table(nets[[net]]$to))] + table(nets[[net]]$to)/length(nets)
  }
  df <- data.frame(genes = genes, average_degree = g[genes],
                   label = annot[match(genes, rownames(annot)), "label"])%>%
    mutate(pwm = ifelse(genes %in% known_tfs, "Available", "Missing"))
  
  df %>%
    filter(genes %in% tfs) %>%
    ggplot(aes(x=average_degree, fill = pwm)) + 
    geom_histogram()+facet_wrap(~ pwm)+ theme_bw()+
    ggtitle(title) + theme(legend.position = "none")
}
(get_degree_hist(rf_nets_0, "weightedRF alpha 0")+
  get_degree_hist(rf_nets_1, "weightedRF alpha 1")+
  get_degree_hist(rf_nets, "weightedRF alpha specific"))/
  (get_degree_hist(lasso_nets_0, "weightedLASSO alpha 0")+
  get_degree_hist(lasso_nets_1, "weightedLASSO alpha 1")+
  get_degree_hist(lasso_nets, "weightedLASSO alpha specific"))

```


```{r, eval=FALSE}
# TGA4
DIANE::get_gene_information(net_rf[net_rf$from == "AT5G10030",]$to, organism = "Arabidopsis thaliana")
DIANE::get_gene_information(net_lasso[net_lasso$from == "AT5G10030",]$to, organism = "Arabidopsis thaliana")

#TGA1
DIANE::get_gene_information(net_rf[net_rf$from == "AT5G65210",]$to, organism = "Arabidopsis thaliana")
DIANE::get_gene_information(net_lasso[net_lasso$from == "AT5G65210",]$to, organism = "Arabidopsis thaliana")


# HH02
DIANE::get_gene_information(net_rf[net_rf$from == "AT1G68670",]$to, organism = "Arabidopsis thaliana")
DIANE::get_gene_information(net_lasso[net_lasso$from == "AT1G68670",]$to, organism = "Arabidopsis thaliana")

get_in_hubs <- function(net){
  hubs <- names(table(net$to)[order(table(net$to), decreasing = T)])[1:25]
  DIANE::get_gene_information(hubs, organism = "Arabidopsis thaliana")
}
get_in_hubs(net_lasso)
get_in_hubs(net_rf)

DIANE::get_gene_information(net_rf[net_rf$to == "AT1G08090",]$from, organism = "Arabidopsis thaliana")

```


```{r}

# load("results/alpha_per_gene_weighted_LASSO_test.rdata")
# load("results/alpha_per_gene_weighted_RF.rdata")
# 

load("rdata/pwm_prom_jaspar_dap.rdata")
load("rdata/gene_structure.rdata")


# load("results/gene_specific_grns.rdata")
# load("results/gene_specific_mse.rdata")
# load("results/gene_specific_edges.rdata")

mean_expr <- rowMeans(counts)[genes]
var_expr <- matrixStats::rowSds(counts[genes,])*matrixStats::rowSds(counts[genes,])
pwm_prom_n_TFs <- pwm_prom[pwm_prom$TF %in% tfs,]


# to comment for new version where mse is already normalized per genes
# norm_mse <- exp(as.matrix(cbind(lmses, lmses_lasso)))/var_expr
genes_info <- data.frame(genes = genes,
                         cluster_rf = alphas_rf,
                         cluster_lasso = alphas_lasso)

genes_info$is_tf <- genes %in% tfs
genes_info$var <- var_expr
genes_info$expr <- mean_expr
genes_info$nb_motifs <- table(pwm_prom$target)[genes]
genes_info$nb_motifs_n_tfs <- table(pwm_prom_n_TFs$target)[genes]
genes_info$common <- genes %in% common

genes_info[,c("n_introns", "n_transcripts")] <-
  gene_structure[match(genes_info$gene, gene_structure$gene),
                 c("n_introns", "n_transcripts")]


# for rf
genes_info%>%
  ggplot(aes(x=factor(cluster_rf), y=log(n_introns))) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of introns for RF groups")) +
genes_info%>%
  ggplot(aes(x=factor(cluster_rf), y=n_transcripts)) +
  geom_boxplot(width=0.1, fill = "white")+
  geom_violin(fill="darkblue", alpha=0.2)+
  ggtitle(("Number of transcripts for RF groups")) +
  stat_compare_means()

genes_info%>%
  ggplot(aes(x=factor(cluster_rf), y=nb_motifs)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of motifs in promoter for RF groups")) +
  genes_info%>%
  ggplot(aes(x=factor(cluster_rf), y=nb_motifs_n_tfs)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of motifs of N-responsive TFs in promoter for RF groups")) +
  stat_compare_means() 


genes_info%>%
  ggplot(aes(x=factor(cluster_rf), y=var)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Gene variance for RF groups")) + scale_y_log10()+
  stat_compare_means()+ genes_info%>%
  ggplot(aes(x=factor(cluster_rf), y=expr)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Gene expression for RF groups")) + scale_y_log10()+
  stat_compare_means()


genes_info %>%
  group_by(cluster_rf) %>%
  summarise(n=n(),
            tf_frac=sum(is_tf)/n()) %>%
  ggplot(aes(x=cluster_rf, y=tf_frac,
                          label = paste("n=",n))) +
  geom_bar(stat = "identity", aes(fill=tf_frac), alpha = 1)+
  geom_hline(yintercept = length(tfs)/length(genes)) +
  geom_text(y=0.2) + xlab("cluster RF") +
  ggtitle("Fraction of TFs in RF groups")+ylim(c(0,0.2))





###########" for lasso

genes_info%>%
  ggplot(aes(x=factor(cluster_lasso), y=log(n_introns))) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of introns for LASSO groups")) +
genes_info%>%
  ggplot(aes(x=factor(cluster_lasso), y=n_transcripts)) +
  geom_boxplot(width=0.1, fill = "white")+
  geom_violin(fill="darkblue", alpha=0.2)+
  ggtitle(("Number of transcripts for LASSO groups")) +
  stat_compare_means()

genes_info%>%
  ggplot(aes(x=factor(cluster_lasso), y=nb_motifs)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of motifs in promoter for LASSO groups")) +
  genes_info%>%
  ggplot(aes(x=factor(cluster_lasso), y=nb_motifs_n_tfs)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of motifs of N-responsive TFs in promoter for LASSO groups")) +
  stat_compare_means() 


genes_info%>%
  ggplot(aes(x=factor(cluster_lasso), y=var)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Gene variance for LASSO groups")) + scale_y_log10()+
  stat_compare_means()+ genes_info%>%
  ggplot(aes(x=factor(cluster_lasso), y=expr)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Gene expression for LASSO groups")) + scale_y_log10()+
  stat_compare_means()


genes_info %>%
  group_by(cluster_lasso) %>%
  summarise(n=n(),
            tf_frac=sum(is_tf)/n()) %>%
  ggplot(aes(x=cluster_lasso, y=tf_frac,
                          label = paste("n=",n))) +
  geom_bar(stat = "identity", aes(fill=tf_frac), alpha = 1)+
  geom_hline(yintercept = length(tfs)/length(genes)) +
  geom_text(y=0.2) + xlab("cluster RF") +
  ggtitle("Fraction of TFs in RF groups")+ylim(c(0,0.2))

######## common classes of interest as compared to the rest
length(intersect(common, tfs))/length(common)


genes_info%>%
  ggplot(aes(x=common, y=log(n_introns))) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of introns for common genes of interest")) +
genes_info%>%
  ggplot(aes(x=common, y=n_transcripts)) +
  geom_boxplot(width=0.1, fill = "white")+
  geom_violin(fill="darkblue", alpha=0.2)+
  ggtitle(("Number of transcripts for common groups")) +
  stat_compare_means()

genes_info%>%
  ggplot(aes(x=common, y=nb_motifs)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of motifs in promoter for common groups")) +
  genes_info%>%
  ggplot(aes(x=common, y=nb_motifs_n_tfs)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Number of motifs of N-responsive TFs in promoter for common groups")) +
  stat_compare_means() 


genes_info%>%
  ggplot(aes(x=common, y=var)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Gene variance for common groups")) + scale_y_log10()+
  stat_compare_means()+ genes_info%>%
  ggplot(aes(x=common, y=expr)) +
  geom_violin(fill="darkblue", alpha=0.2)+geom_jitter(width=0.1)+
  geom_boxplot(width=0.1, fill = "white")+
  ggtitle(("Gene expression for common groups")) + scale_y_log10()+
  stat_compare_means()


genes_info %>%
  group_by(common) %>%
  summarise(n=n(),
            tf_frac=sum(is_tf)/n()) %>%
  ggplot(aes(x=common, y=tf_frac,
                          label = paste("n=",n))) +
  geom_bar(stat = "identity", aes(fill=tf_frac), alpha = 1)+
  geom_hline(yintercept = length(tfs)/length(genes)) +
  geom_text(y=0.2) + xlab("cluster common") +
  ggtitle("Fraction of TFs in common groups")+ylim(c(0,0.22))


phyper(q=length(intersect(common, tfs)), 
       m = length(tfs), 
       n = length(genes) - length(tfs),
       k = length(common),lower.tail = F)
```




<!-- ```{r} -->
<!-- # promoteurs enrichis en certains motifs de TFs? -->
<!-- known_tfs <- tfs[which(tfs %in% pwm_prom$TF)] -->

<!-- get_number_of_motifs_per_tfs <- function(genes){ -->
<!--   table(pwm_prom[pwm_prom$target %in% genes & pwm_prom$TF %in% tfs,"TF"])[known_tfs] -->
<!-- } -->

<!-- targets_per_pwm <- data.frame(row.names = known_tfs) -->
<!-- for(group in unique(clusters_rf)){ -->
<!--   targets_per_pwm[paste("lasso", group)] <- get_number_of_motifs_per_tfs(names( -->
<!--     which(clusters_lasso == group)))/sum(clusters_lasso == group) -->
<!--   targets_per_pwm[paste("rf", group)] <- get_number_of_motifs_per_tfs(names( -->
<!--     which(clusters_rf == group)))/sum(clusters_rf == group) -->

<!-- } -->

<!-- enrichments_per_pwm <- data.frame(row.names = known_tfs) -->
<!-- n_genes <- length(genes) -->
<!-- for(group in unique(clusters_rf)){ -->
<!--   # number of motifs in all the genes -->
<!--   n_targets_lasso_in_all <- get_number_of_motifs_per_tfs(genes) -->

<!--   n_targets_lasso_in_group <- get_number_of_motifs_per_tfs(names( -->
<!--     which(clusters_lasso == group))) -->
<!--   n_group_lasso <- length(names(which(clusters_lasso == group))) -->

<!--   n_targets_rf_in_group <- get_number_of_motifs_per_tfs(names( -->
<!--     which(clusters_rf == group))) -->
<!--   n_group_rf <- length(names(which(clusters_rf == group))) -->

<!--   for(tf in known_tfs){ -->

<!--     # number of genes with that motif in all genes -->
<!--     n_targets_in_all_tf <- n_targets_lasso_in_all[tf] -->

<!--     # number of genes with that motif in the lasso group -->
<!--     n_targets_lasso_in_group_tf <- n_targets_lasso_in_group[tf] -->
<!--     p_lasso <- phyper(q=n_targets_lasso_in_group_tf-1, -->
<!--            m=n_targets_in_all_tf, #white balls -->
<!--            n=n_genes-n_targets_in_all_tf, # black balls -->
<!--            k=n_group_lasso, lower.tail = FALSE) -->

<!--      # number of genes with that motif in the rf group -->
<!--     n_targets_rf_in_group_tf <- n_targets_rf_in_group[tf] -->
<!--     p_rf <- phyper(q=n_targets_rf_in_group_tf-1, -->
<!--            m=n_targets_in_all_tf,  -->
<!--            n=n_genes-n_targets_in_all_tf, -->
<!--            k=n_group_rf, lower.tail = FALSE) -->

<!--     enrichments_per_pwm[tf, paste0(group, "lasso")]<- p_lasso -->
<!--     enrichments_per_pwm[tf, paste0(group, "rf")]<- p_rf -->
<!--   } -->
<!-- } -->

<!-- enrichments_per_pwm[enrichments_per_pwm<0.05] <- 0 -->
<!-- enrichments_per_pwm[enrichments_per_pwm>=0.05] <- 1 -->
<!-- Heatmap(enrichments_per_pwm, cluster_columns = F) -->


<!-- tfs_rf_pwm_pos <- rownames(enrichments_per_pwm[enrichments_per_pwm$`2rf`==0,]) -->
<!-- tfs_lasso_pwm_pos <- rownames(enrichments_per_pwm[enrichments_per_pwm$`1lasso`==0,]) -->
<!-- DIANE::get_gene_information(tfs_rf_pwm_pos, organism = "Arabidopsis thaliana") -->
<!-- DIANE::get_gene_information(tfs_lasso_pwm_pos, organism = "Arabidopsis thaliana") -->
<!-- DIANE::get_gene_information(intersect(tfs_rf_pwm_pos, tfs_lasso_pwm_pos ), organism = "Arabidopsis thaliana") -->

<!-- tfs_rf_pwm_bad <- rownames(enrichments_per_pwm[enrichments_per_pwm$`1rf`==0,]) -->
<!-- tfs_lasso_pwm_bad <- rownames(enrichments_per_pwm[enrichments_per_pwm$`2lasso`==0,]) -->
<!-- DIANE::get_gene_information(tfs_rf_pwm_bad, organism = "Arabidopsis thaliana") -->
<!-- DIANE::get_gene_information(tfs_lasso_pwm_bad, organism = "Arabidopsis thaliana") -->

<!-- tfs_rf_pwm_pretty_bad <- rownames(enrichments_per_pwm[enrichments_per_pwm$`no diffrf`==0,]) -->
<!-- tfs_lasso_pwm_pretty_bad <- rownames(enrichments_per_pwm[enrichments_per_pwm$`no difflasso`==0,]) -->
<!-- DIANE::get_gene_information(tfs_rf_pwm_pretty_bad, organism = "Arabidopsis thaliana") -->
<!-- DIANE::get_gene_information(tfs_lasso_pwm_pretty_bad, organism = "Arabidopsis thaliana") -->
<!-- ``` -->


<!-- # Network ranks of the Tfs that have enriched motifs in some gene clusters -->



<!-- ```{r, fig.width=10, fig.height=10, eval=FALSE} -->
<!-- load("results/100_permutations_bRF_edges.rdata") -->
<!-- # 10 first replicates -->
<!-- edges <- edges[names(edges)[as.numeric(str_split_fixed(names(edges), '_', 5)[,4])<=10]] -->
<!-- net <- edges$bRF_1_trueData_1_0.005 -->
<!-- # nodes <-  -->
<!-- get_nodes_relative_rank <- function(net, nodes){ -->
<!--   return(setNames(rank(table(net$from))[nodes]/length(unique(net$from)),nodes)) -->
<!-- } -->

<!-- positive_genes <- names(clusters_rf[clusters_rf==2]) -->
<!-- edges_positives <- lapply(edges, function(net){net[net$to %in% positive_genes,]}) -->

<!-- negative_genes <- names(clusters_rf[clusters_rf==1]) -->
<!-- edges_negatives <- lapply(edges, function(net){net[net$to %in% negative_genes,]}) -->


<!-- relative_rank <- lapply(edges, get_nodes_relative_rank, tfs_rf_pwm_pos) -->
<!-- relative_rank_pos <- lapply(edges_positives, get_nodes_relative_rank, tfs_rf_pwm_pos) -->

<!-- data.frame(relative_rank) %>% -->
<!--   rownames_to_column("gene") %>% -->
<!--   reshape2::melt() %>% -->
<!--   separate(variable, into = c("method", "alpha", "dataset", "rep", "density"), sep = '_', remove = F) %>% -->
<!--   mutate(alpha = as.numeric(alpha)) %>% -->
<!--   ggplot(aes(x=alpha, y=value, color = dataset)) +  -->
<!--   ggh4x::facet_nested_wrap(vars(density, gene), nest_line = T) +  -->
<!--   geom_point() + geom_smooth() + -->
<!--   theme(strip.background = element_blank(), axis.title.x = element_text(size = 22), -->
<!--         title = element_text(size = 16), strip.text = element_text(size = 16),  -->
<!--         legend.text = element_text(size = 15), axis.text = element_text(size = 12),  -->
<!--         legend.position = 'left') + ggtitle("TFs with motifs enriched in positive genes promoters : relative out-degree rank") -->


<!-- data.frame(relative_rank_pos) %>% -->
<!--   rownames_to_column("gene") %>% -->
<!--   reshape2::melt() %>% -->
<!--   separate(variable, into = c("method", "alpha", "dataset", "rep", "density"), sep = '_', remove = F) %>% -->
<!--   mutate(alpha = as.numeric(alpha)) %>% -->
<!--   ggplot(aes(x=alpha, y=value, color = dataset)) +  -->
<!--   ggh4x::facet_nested_wrap(vars(density, gene), nest_line = T) +  -->
<!--   geom_point() + geom_smooth() + -->
<!--   theme(strip.background = element_blank(), axis.title.x = element_text(size = 22), -->
<!--         title = element_text(size = 16), strip.text = element_text(size = 16),  -->
<!--         legend.text = element_text(size = 15), axis.text = element_text(size = 12),  -->
<!--         legend.position = 'left') + ggtitle("TFs with motifs enriched in positive genes promoters : relative out-degree rank in positive subset") -->



<!-- relative_rank_bad <- lapply(edges, get_nodes_relative_rank, tfs_rf_pwm_bad) -->
<!-- relative_rank_bad_subset <- lapply(edges_negatives, get_nodes_relative_rank, tfs_rf_pwm_bad) -->

<!-- data.frame(relative_rank_bad) %>% -->
<!--   rownames_to_column("gene") %>% -->
<!--   reshape2::melt() %>% -->
<!--   separate(variable, into = c("method", "alpha", "dataset", "rep", "density"), sep = '_', remove = F) %>% -->
<!--   mutate(alpha = as.numeric(alpha)) %>% -->
<!--   ggplot(aes(x=alpha, y=value, color = dataset)) +  -->
<!--   ggh4x::facet_nested_wrap(vars(density, gene), nest_line = T) +  -->
<!--   geom_point() + geom_smooth() + -->
<!--   theme(strip.background = element_blank(), axis.title.x = element_text(size = 22), -->
<!--         title = element_text(size = 16), strip.text = element_text(size = 16),  -->
<!--         legend.text = element_text(size = 15), axis.text = element_text(size = 12),  -->
<!--         legend.position = 'left') + ggtitle("TFs with motifs enriched in negative genes promoters : relative out-degree rank") -->

<!-- data.frame(relative_rank_bad_subset) %>% -->
<!--   rownames_to_column("gene") %>% -->
<!--   reshape2::melt() %>% -->
<!--   separate(variable, into = c("method", "alpha", "dataset", "rep", "density"), sep = '_', remove = F) %>% -->
<!--   mutate(alpha = as.numeric(alpha)) %>% -->
<!--   ggplot(aes(x=alpha, y=value, color = dataset)) +  -->
<!--   ggh4x::facet_nested_wrap(vars(density, gene), nest_line = T) +  -->
<!--   geom_point() + geom_smooth() + -->
<!--   theme(strip.background = element_blank(), axis.title.x = element_text(size = 22), -->
<!--         title = element_text(size = 16), strip.text = element_text(size = 16),  -->
<!--         legend.text = element_text(size = 15), axis.text = element_text(size = 12),  -->
<!--         legend.position = 'left') + ggtitle("TFs with motifs enriched in negative genes promoters : relative out-degree rank in negative subset") -->

<!-- ``` -->



<!-- # Go enrichments -->

<!-- ```{r, eval=T} -->
<!-- library(DIANE) -->
<!-- background <- convert_from_agi(genes) -->

<!-- for(group in unique(clusters_rf)){ -->

<!--   genes_i <- names(which(clusters_lasso == group)) -->

<!--   print(paste("LASSO", length(genes_i), "genes,", group, "\n")) -->
<!--   genes_i <- convert_from_agi(genes_i) -->
<!--   go_lasso <- enrich_go(genes_i, background) -->
<!--   DIANE::draw_enrich_go(go_lasso) -->
<!--   print(go_lasso) -->
<!--   genes_i <- names(which(clusters_rf == group)) -->
<!--   print(paste("RF", length(genes_i), "genes, group", group)) -->
<!--   genes_i <- convert_from_agi(genes_i) -->
<!--   go_rf <- enrich_go(genes_i, background) -->
<!--   DIANE::draw_enrich_go(go_rf) -->
<!--   print(go_rf) -->
<!-- } 

-->



<!-- ``` -->

<!-- # Expression -->

<!-- ```{r} -->
<!-- expression <- data.frame(counts) -->
<!-- expression <- (expression-rowMeans(expression)) / matrixStats::rowSds(as.matrix(expression)) -->

<!-- Heatmap(expression, cluster_columns = F, show_row_names = F)+ -->
<!--   rowAnnotation( -->
<!--     clusters_rf = clusters_rf[rownames(expression)], -->
<!--     clusters_lasso = clusters_lasso[rownames(true_mse)], -->
<!--     col=list(clusters_lasso= setNames(c( "darkgreen","darkorange", "lightgrey"),  -->
<!--                                        nm = names(table(clusters_lasso))), -->
<!--       clusters_rf= setNames(c("darkorange", "darkgreen", "lightgrey"),  -->
<!--                                        nm = names(table(clusters_rf))))) -->
<!-- ``` -->


<!-- # intersection of groups -->

<!-- ```{r} -->
<!-- load("results/clusters_mse_bRF_100permutations.rdata") -->
<!-- load("results/clusters_mse_lasso_100permutations.rdata") -->
<!-- DIANE::draw_venn(list("positive for lasso" = names(clusters_lasso[clusters_lasso==1]), -->
<!--                       "positive for RFs" = names(clusters_rf[clusters_rf==2]))) -->

<!-- DIANE::draw_venn(list("negative for lasso" = names(clusters_lasso[clusters_lasso==2]), -->
<!--                       "negative for RFs" = names(clusters_rf[clusters_rf==1]))) -->

<!-- DIANE::draw_venn(list("no diff for lasso" = names(clusters_lasso[clusters_lasso=="no diff"]), -->
<!--                       "no diff for RFs" = names(clusters_rf[clusters_rf=="no diff"]))) -->
<!-- ``` -->






